# =======================================================================================================
# This is a script for generating Figure4: Allele specific expression across rare disease samples
# #
# input:
#       1. overviewstats.tsv, numASEoutliers_perRDSind.txt, amelieScore_randomVsMatched_data.RData, deleteriouspv_ase_forscatterdensity.RData
#
# output:
#       1. Figure4A.pdf, Figure4B.pdf, Figure4C.pdf, Figure4.pdf
#       2. Figure4.out.RData
#          Store all data used and generated by running this script
#
# Note that the final figure was generated by using inkscape by combining figures from this script for visualization purposes
#
# =======================================================================================================



#!/bin/R


library(data.table)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(reshape2)
library(cowplot)
library(ggrepel)
library(readr)
library(ggpubr)


rm(list=ls())



# Master directory
dir = Sys.getenv('RARE_DIS_DIR')



fsize=15
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize-1),
	axis.text.y= element_text(size=fsize-1), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.1),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=9),
	panel.border=element_blank()) 


# Figure 4A
df<-read.table(paste(dir,"/data/ase/June13/hetgenes/overviewstats.tsv", sep=""), header=T)

outliers=read_tsv(paste(dir,"/data/rds_data/numASEoutliers_perRDSind.txt", sep="")
colnames(outliers)[1]<-"SampID"
colnames(outliers)[2]<-"ASE outlier"

df<-merge(df,outliers)
df<-df[which(df$in_freeze=="yes"),]

df[is.na(df)] <- 0
drops <- c("UniqueASE","AI.1", "reads20", "reads20_not01", "AI.80")
df<-df[ , !(names(df) %in% drops)]



meltData<-melt(df, id=c("affected_status","SampID", "variant_data", "in_freeze"))
color_variant=c( "#2F4E6F", "#98B1C4", "#C8D7E3")

nplot = ggplot(meltData, aes(x=factor(variable), y=as.numeric(value), Group=variant_data)) +
  geom_boxplot(aes(fill=variant_data)) + xlab('') +
  ylab('Number per sample') +
 RD_theme+ scale_fill_manual(values=color_variant,name="Variant data")+
 scale_y_log10()+ 
 theme(legend.position=c(0.8,0.9))+
 scale_x_discrete(labels=c("allhets" = "Heterozygous\nsites",  "AI.65"="Allelic ratio\nover 0.65" ))

ggsave('Figure4A.pdf', nplot, path=paste(dir,"/analysis/manuscript/figures/", sep=""), width=6, height=6)



# Figure 4.B
cat.colors = c('grey', '#DEADA1')
data_dir = '/users/nferraro/data/rds_data/'

load(paste0(dir, '/data/rds_data/amelieScore_randomVsMatched_data.RData'))


score_plot=ggplot(allScores, aes(x=HPO_terms, y=PropNonZero, fill=HPO_terms)) + 
  geom_violin(trim=TRUE)+
  geom_boxplot(width=0.1, fill="white")+
  labs(x="", y = "Proportion of genes\n with Amelie score > 0 per sample") + RD_theme+ stat_compare_means(label.x = 1.5,method.args = list(alternative ="greater"),aes(label = paste0("p = ", ..p.format..))) + scale_fill_manual(values=cat.colors)+theme(legend.position="")

ggsave('Figure4B.pdf', score_plot, path=paste(dir,"/analysis/manuscript/figures/", sep=""), width=6, height=6)


# Figure 4C
load(paste0(dir,"data/ase/deleteriouspv_ase_forscatterdensity.RData"))


pv_ase$mytext <- ""
pv_ase[which(pv_ase$indID=="RD046" & pv_ase$ensgene =="ENSG00000142634" & pv_ase$classification2=="xstop"),]$mytext="EFHD2"

pv_ase_plot2<-ggplot(pv_ase%>%arrange(classification2), aes(x=TotalReads+1, y=RefRatio, color=classification2))+ geom_point(size=3,alpha=0.5)+
  geom_hline(yintercept=0.65, linetype="dashed", color = "black", size=1.5) +
  geom_hline(yintercept=0.35, linetype="dashed", color = "black", size=1.5) +
  scale_color_manual(values=c("#C5C5C5", "#a8caed","#eda8a8","#368ce2","#e83333"), name="",breaks=c("rare", "splice", "stop", "xsplice", "xstop"),labels=c("Rare variant", "Rare splicing variant", "Rare stop variant","Rare splicing variant\n biased alternative","Rare stop variant\nbiased alternative")) + 
  scale_size_manual(values=c(3,3,3,3,3)) +
  scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x))) +RD_theme +
  theme(legend.position="top")+
  geom_text_repel(data=pv_ase, aes(label = mytext), color="black", size = 4,  box.padding = unit(0.35, "lines"), point.padding = unit(0.3, "lines"),nudge_x= 0.2)+
  labs(x="Total number of reads covering the variant +1\n(log10 scale)", y="Proportion of reference\nallele")


pv_density_plot<-ggplot(pv_ase_density%>%arrange(classification), aes(x=RefRatio,color=classification))+ 
geom_density(alpha=0.5) +
scale_color_manual(values=c("#C5C5C5", "#a8caed","#eda8a8")) +
coord_flip() +
theme(legend.position="none", axis.title.y=element_blank(), axis.text.y=element_blank())



figure4C=ggdraw()+draw_plot(pv_ase_plot2, 0,0,3/4,1)+ draw_plot(pv_density_plot, 3/4,0,1/4,1)

ggsave('Figure4C.pdf', figure4C, path=paste(dir,"/analysis/manuscript/figures/", sep=""), width=6, height=6)


# entire figure
figure4_plot=ggdraw()+draw_plot(nplot, 0,2/3,1/2,1/3)+
	draw_plot(score_plot, 1/2,2/3,1/2,1/3)+
	draw_plot(pv_ase_plot2, 0,0,3/4,2/3)+
	draw_plot(pv_density_plot, 3/4,0,1/4,2/3)+
	draw_plot_label(c('A', 'B', 'C'), c(0,1/2,0), c(1,1,2/3), size = 20)

pdf(paste(dir,"/analysis/manuscript/figures/Figure4.pdf", sep=""), w=12, h=12)
figure4_plot
dev.off()


# Save data
save.image(file = paste(dir,"/data/Figure4.out.RData",sep=""))
