#!/bin/R


rm(list=ls())


# Master directory
dir = Sys.getenv('RARE_DIS_DIR')
## FUNCTIONS

source(paste(dir,"/scripts/manuscript_analyses/Figures_source.R", sep=""))




#Libraries
library(ggplot2)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(sva)
library(cowplot)
library(purrr)
library(broom)
library(reshape2)
library(ggfortify)
library(gridExtra)
library(data.table)
library(annotables)
library(corrplot)
library(RColorBrewer)
library(qvalue) 
library(ggpubr)
library(missMDA)
library(parallel)
library(foreach)
require(doMC)

#--- MAIN
# Get metadata
metadata = paste(dir,"/data/metadata/2018_12_02_Rare_Disease_Metadata.tsv",sep="")
metadata = read_tsv(metadata) %>% filter(in_freeze=="yes") %>% filter(source_of_RNA=="Blood", status=="PASSED")

# Get metadata info for pivus and dgn
meta_pivus=paste(dir,"/data/metadata/PIVUS_RNASequencingInfo.csv",sep="")
meta_pivus=read_csv(meta_pivus) %>% filter(Age=="70") %>% filter(RunOK=="Yes")
meta_pivus2=paste(dir,"/data/metadata/Pivus_expCovariatesAll_unscaled.txt",sep="")
meta_pivus2=read_tsv(meta_pivus2)%>% filter(Age==0) 

# Read in annotation files linking HPO terms genes and phenotypes
# generate HPO terms list for each sample
gene_to_pheno=paste0(dir,"/data/metadata/18_10_23_ALL_SOURCES_ALL_FREQUENCIES_genes_to_phenotype.txt")
gene_to_pheno = read_tsv(gene_to_pheno, comment= "#",col_names = FALSE)
colnames(gene_to_pheno)=c("entrez", "symbol", "DiseaseId", "HPO_ID")

pheno_annot=paste0(dir,"/data/metadata/18_10_23_ALL_SOURCES_ALL_FREQUENCIES_phenotype_to_genes.txt")
pheno_annot=read_tsv(pheno_annot, comment= "#",col_names = FALSE)
colnames(pheno_annot)=c("HPO_ID", "HPO_name", "entrez", "symbol")

# Read in and process alternative HPO files (generated by parse_hpoterms.py)
parent_terms=paste0(dir,"/data/metadata/parent_terms_hpo_2018_10_23.txt")
child_terms=paste0(dir,"/data/metadata/child_terms_hpo_2018_10_23.txt")
altid_terms=paste0(dir,"/data/metadata/altid_terms_hpo_2018_10_23.txt")

altid_terms=as.data.frame(read.table(altid_terms, header=F, sep="\t"))
colnames(altid_terms)=c("HPO_ID", "ALT_ID")
child_terms=as.data.frame(read.table(child_terms, header=F, sep="\t"))
colnames(child_terms)=c("HPO_ID", "ALT_ID")
parent_terms=as.data.frame(read.table(parent_terms, header=F, sep="\t"))
colnames(parent_terms)=c("HPO_ID", "ALT_ID")


# generates list of HPO alternative IDs from the input files
alt_hpo_ids_list=lapply(altid_terms$HPO_ID,get_hpo_alt_id, altid_terms=altid_terms)
names(alt_hpo_ids_list)=altid_terms$HPO_ID
child_hpo_ids_list=lapply(child_terms$HPO_ID,get_hpo_alt_id, altid_terms=child_terms)
names(child_hpo_ids_list)=child_terms$HPO_ID
parent_hpo_ids_list=lapply(parent_terms$HPO_ID,get_hpo_alt_id, altid_terms=parent_terms)
names(parent_hpo_ids_list)=parent_terms$HPO_ID

# For every ID associated with the disease, get list of alternative Ids, child ids and parent ids
# make a list for each case with HPO terms

cases=metadata %>% filter(affected_status=="Case") %>% select(sample_id) %>% pull

# Get list of HPO terms from metadata file
HPOs=lapply(cases,get_list_hpo, metadata)
names(HPOs)=cases

# select cases with genetic data
cases_withGen=metadata %>% filter(variant_data !="none", affected_status=="Case")%>% select(sample_id) %>% pull
cases_withGen=cases_withGen[!cases_withGen=="RD008"]

# Generate extended list of HPO terms for each sample
HPO_extended=lapply(cases, process_sample_ids_hpo, HPO_list=HPOs)
names(HPO_extended)=cases

HPO_extended <- Map(cbind, HPO_extended, sample = names(HPO_extended))
HPO_extended.df=as.data.frame(do.call("rbind", HPO_extended))
colnames(HPO_extended.df)=c("HPO_term", "sample")
HPO_extended.df$sample=as.character(HPO_extended.df$sample)


# attach gene to HPO ids for each samples
HPO_extended.df=HPO_extended.df %>% left_join(pheno_annot, by=c("HPO_term"="HPO_ID"), all.left=T) %>% left_join(grch37, by="symbol", all.left=T)  %>% select(HPO_term,sample, ensgene) %>% unique %>% filter(!is.na(ensgene))


# get affected status data
affected_status_PIVUS=data.frame(sample_id=meta_pivus2$Sample, affected_status=rep("Control", nrow(meta_pivus2)))

affected_status_df=rbind(metadata%>% select(sample_id,affected_status),affected_status_PIVUS)
colnames(affected_status_df)=c('sample', 'status')

affected_status_df$cohort=c(rep("RD",nrow(metadata)), rep("PIVUS",  nrow(meta_pivus2)))

samples=affected_status_df$sample
RD_samples=metadata$sample_id

# select samples for which we have genetic data
sample_with_data=metadata %>% filter(variant_data=="exome" | variant_data=="genome")%>% select(sample_id) %>% pull
sample_with_data=sample_with_data[!sample_with_data=="RD008"]


# read in exac data
exac = as.data.frame(read.table('/users/lfresard/NumberOfIndividuals_Impact/enrichment_disease-conservation_database/annotations/EXAC/forweb_cleaned_exac_r03_march16_z_data_pLI.txt', sep = '\t', stringsAsFactors = F, header = T))
exac=exac %>% inner_join(grch37, by=c("gene"="symbol")) %>% select(ensgene,syn_z, mis_z, lof_z, pLI) 



# annotation of rare variants

rv_annotation="/users/xli6/projects/udn/feature_file/udn_variant_annotation.txt"
rv_annotation=read_tsv(rv_annotation)
rv_annotation=rv_annotation %>% mutate(gene=str_extract(gene_id, "ENSG[0-9]+"), variantPos=pos-1)


#get splicing z-scores
zscores_blood =as.data.frame(read.table(paste(dir, "/analysis/outlier_analysis/splicing/2018_12_3_paper_revisions/2018_12_03_splicing_zscores.txt", sep=""), sep="\t", header=T))

zscores_blood =zscores_blood %>% mutate(gene=unlist(strsplit(as.character(gene),"[.]"))[c(TRUE,FALSE)]) 
genes=zscores_blood %>% select(gene) %>% distinct %>% pull

#get splicing z-scores filtered with rare variants data

RV_outlier_file=paste(dir, "/analysis/outlier_analysis/splicing/2018_12_3_paper_revisions/RV/2018_12_03_splicing_zscores_sorted_RV_window.txt", sep="")
RV_outlier=as.data.frame(read.table(RV_outlier_file, sep="\t", header=F))
colnames(RV_outlier)=c(colnames(zscores_blood), "chr_var", "start_var", "end_var", "ref_var", "alt_var", "gnomAD_AF", "raw_cadd", "phred_cadd", "chr_gene", "start_gene", "end_gene", "gene_name")

# make a column with singleton included in the analysis
RV_outlier =RV_outlier %>% mutate(gnomAD_AF_withsgl = replace(gnomAD_AF, gnomAD_AF==".", 0))


# format variant data so that only the max observed allele frequency is kept for variant were several frequencies are available
RV_outlier$gnomAD_AF_filt_withsgl=sapply(RV_outlier$gnomAD_AF_withsgl,process_variant)

#filter for RV with MAF<=0.1%
maf=0.001
RV_outlier_filt_withsgl=RV_outlier %>% filter(!is.na(gnomAD_AF_filt_withsgl)) %>% filter(gnomAD_AF_filt_withsgl <=maf)
RV_outlier_filt_withsgl =RV_outlier_filt_withsgl %>% mutate(gene_name=str_extract(gene_name, "ENSG[0-9]+"))

RV_outlier_filt_withsgl_annot=rv_annotation %>% left_join(RV_outlier_filt_withsgl, by=c("gene","variantPos"="start_var", "indiv_id"="variable"))


# Make dataframe containing the number of genes with at least 1 outlier at different Zscore thresholds
sample_outlier.df=data.frame(sample_id=samples,
	"Z2"=sapply(samples, get_outlier_genes, threshold=2, outlier_df.m=zscores_blood, 0),
	"Z3"=sapply(samples, get_outlier_genes, threshold=3, outlier_df.m=zscores_blood, 0),
	"Z4"=sapply(samples, get_outlier_genes, threshold=4, outlier_df.m=zscores_blood, 0),
	"Z5"=sapply(samples, get_outlier_genes, threshold=5, outlier_df.m=zscores_blood, 0),
	"Z6"=sapply(samples, get_outlier_genes, threshold=6, outlier_df.m=zscores_blood, 0),
	affected_status=sapply(samples,get_affected_status, affected_status_df=affected_status_df))

# Add cohort information to data frame
sample_outlier.df=sample_outlier.df %>% left_join(affected_status_df, by=c("sample_id"="sample")) %>% select(-status) 

# transform data for plotting
sample_outlier.df.m=melt(sample_outlier.df,id.vars=c("sample_id","affected_status", "cohort"))
sample_outlier.df.m$variable=c(rep(2,length(samples)),rep(3,length(samples)),rep(4,length(samples)), rep(5,length(samples)), rep(6,length(samples)))


# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# Figure 3C
# number of RV
RV_junc_dir=paste(dir, "/analysis/outlier_analysis/splicing/2018_12_3_paper_revisions/RV/",sep="")
RV_junc_samples=list.files(pattern='_junction_RV_window.txt', path=RV_junc_dir)


bed_dir=paste(dir, "/data/bed/all_samples/", sep="")

# count the number of RV nearby annotated junctions for each sample
RV_junction_number=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.01, "yes"))
RV_junction_number=RV_junction_number %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_junction_number=RV_junction_number%>% mutate(filter="RV near junction")%>% mutate(splicing_data="no")%>% mutate(singleton="yes")

# count the number of RV  for each sample
RV_number_withsgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV, bed_dir, 0, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV, bed_dir, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_RV, bed_dir, 0.01, "yes"))
RV_number_withsgl=RV_number_withsgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_number_withsgl=RV_number_withsgl%>% mutate(filter="Rare variant")%>% mutate(splicing_data="no")%>% mutate(singleton="yes")

# data with splicing outliers
RV_number_out_withsgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0,  "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.01, "yes"))
RV_number_out_withsgl=RV_number_out_withsgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)
RV_number_out_withsgl=RV_number_out_withsgl %>% mutate(filter="splicing outlier")%>% mutate(splicing_data="yes")%>% mutate(singleton="yes")




# combine info of RV and splicing outlier
RV_outliers=rbind(RV_junction_number,RV_number_withsgl,RV_number_out_withsgl)

# melt DF for plot
RV_outliers.m=melt(RV_outliers,id.var=c("sample", "variant_data", "affected_status", "filter", "splicing_data", "singleton"))

RV_outliers.m_withsgl=RV_outliers.m %>% filter(singleton=="yes", variable!="RV_number_MAF0")


# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# Figure 3 D
## Influence of filtering on splicing outlier results
zscore_threshold=2
distance_threshold=20
cadd=10
pli=0.9

sample_outlier_filter_withsgl.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=0),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=pli),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb,outlier_file=zscores_blood,HPOs_list=HPO_extended, pLI_thre=0),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl, CADD=0, pLI_thre=0),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl, CADD=0, pLI_thre=pli),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl,CADD=cadd, pLI_thre=0),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0),
	"combined"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))

save("sample_outlier_filter_withsgl.df", file=paste0(dir,"/analysis/outlier_analysis/splicing/2018_12_3_paper_revisions/splicing_filters_genes.RData"))


# make a plot with proportions of candidates 

sample_outlier_filter_withsgl.df= sample_outlier_filter_withsgl.df %>% filter(raw!=0)
sample_outlier_filter_withsgl.df$prop_pLI=sample_outlier_filter_withsgl.df$raw.Lof/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_HPO=sample_outlier_filter_withsgl.df$raw.HPO/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RV=sample_outlier_filter_withsgl.df$RV/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVpLI=sample_outlier_filter_withsgl.df$RV.LoF.intolerant/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_CADD=sample_outlier_filter_withsgl.df$RV.CADD/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVHPO=sample_outlier_filter_withsgl.df$RV.HPO/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVHPOCADD=sample_outlier_filter_withsgl.df$RV.HPO.CADD/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_combined=sample_outlier_filter_withsgl.df$combined/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df=sample_outlier_filter_withsgl.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_HPO,prop_RV,prop_RVpLI,prop_CADD, prop_RVHPO,prop_RVHPOCADD,prop_combined)

#melt DF
sample_outlier_filter_withsgl.df.m=melt(sample_outlier_filter_withsgl.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_outlier_filter_withsgl.df.m.cases=sample_outlier_filter_withsgl.df.m %>% filter(affected_status=="Case")

#figure 3E

# filter for Z-score>=2
raw_genes=lapply(cases_withGen, get_outlier_genes_names, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=0)
names(raw_genes)=cases_withGen
	# remove empty fields
raw_genes=compact(raw_genes)
raw_genes <- Map(cbind, raw_genes, sample = names(raw_genes))
raw_genes.df=as.data.frame(do.call("rbind", raw_genes))
raw_genes.df$hpo=mapply(is_hpo,raw_genes.df$sample,raw_genes.df$V1)
raw_genes_hpo_percent=raw_genes.df %>% group_by(sample) %>% mutate(n_genes=n())%>% mutate(percent_hpo=sum(hpo)/n_genes) %>% select(sample,n_genes,percent_hpo)%>% filter(!duplicated(sample)) 
raw_genes_hpo_percent$method="Z-score>=2"

# Z-score >=2, RV within 20bp junction, CADD score >=10
RV_CADD_genes=lapply(cases_withGen, get_outlier_genes_RV_window_names, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl_annot, CADD=10, pLI_thre=0)
names(RV_CADD_genes)=cases_withGen
RV_CADD_genes=compact(RV_CADD_genes)
RV_CADD_genes <- Map(cbind, RV_CADD_genes, sample = names(RV_CADD_genes))
RV_CADD_genes.df=as.data.frame(do.call("rbind", RV_CADD_genes))
RV_CADD_genes.df$hpo=mapply(is_hpo,RV_CADD_genes.df$sample,RV_CADD_genes.df$V1)
RV_CADD_genes_hpo_percent=RV_CADD_genes.df %>% group_by(sample) %>% mutate(n_genes=n()) %>% mutate(percent_hpo=sum(hpo)/n_genes)%>% select(sample,n_genes,percent_hpo)%>% filter(!duplicated(sample)) 
RV_CADD_genes_hpo_percent$method="Z-score>=2, RV, CADD>=10"

## Look at number of rare variants around genes throughout filters
RV_CADD_HPO_genes.df= RV_CADD_genes.df %>% filter(hpo==1)
RV_CADD_HPO_genes.list=split(RV_CADD_HPO_genes.df, as.character(RV_CADD_HPO_genes.df$sample))
RV_CADD_HPO_genes.list=compact(RV_CADD_HPO_genes.list)

# median number of RV per gene for each sample throughout filters
RV_count=data.frame(sample=sample_with_data,
 outliers=sapply(sample_with_data, get_number_RV_pergene, bed_dir, 0.001, "yes",raw_genes),
 outlier_RV_CADD_HPO_junc=sapply(sample_with_data, get_number_RV_pergene, bed_dir, 0.001, "yes",RV_CADD_HPO_genes.list))

RV_count.m=melt(RV_count,id.var=c("sample"))
RV_count.m.filt=RV_count.m %>% filter(variable != "raw_nonout")

custom_comparisons=list( c("outliers","outlier_RV_CADD_HPO_junc"))

# Figure 3 F

# generates results without RV
candidates_no_gen_ext=lapply(cases,get_hpo_comp_nogen,zscores_blood,metadata, gene_to_pheno,pheno_annot,HPO_extended)
names(candidates_no_gen_ext)=cases

candidates_zscore_hpo_ext=do.call("rbind", candidates_no_gen_ext)



# generate results with RV
potential_can=lapply(cases_withGen,get_hpo_comp_window,RV_outlier_filt_withsgl_annot,metadata, gene_to_pheno,pheno_annot, 10,HPO_extended)
names(potential_can)=cases_withGen

candidates_zscore_hpo_RV=do.call("rbind", potential_can)
candidates_zscore_hpo_RV=candidates_zscore_hpo_RV %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, indv_id, institution)


registerDoMC(cores=40)
res_splicing=foreach(n=1:length(cases_withGen), .combine = rbind) %dopar% {
	get_count_filters_splicing(cases_withGen[n])}

#res_splicing=sapply(cases_withGen,get_count_filters_splicing)
rownames(res_splicing)=cases_withGen
res_splicing.df=as.data.frame(res_splicing)
res_splicing.df$filter=rownames(res_splicing.df)
res_splicing.df=res_splicing.df[-1,]
res_splicing.m=res_splicing.df %>% gather(variable, value, -filter)


res_splicing_number.df=data.frame(sample=as.factor(res_splicing.m$variable), value=as.numeric(res_splicing.m$value), filter=unlist(res_splicing.m$filter))
#exp_outlier_number.df=data.frame(sample=as.factor(exp_outlier_number.m$variable), value=unlist(exp_outlier_number.m$value), filter=unlist(exp_outlier_number.m$filter))
# order factors
res_splicing_number.df$variable=factor(res_splicing_number.df$variable, levels=rev(c("RV","RV_CADD", "RV_HPO","RV_CADD_HPO","RV_JUNC", "RV_JUNC_CADD","RV_JUNC_HPO", "RV_JUNC_CADD_HPO", "HPO" ,"SPLI_OUTLIER", "SPLI_OUTLIER_PLI","SPLI_OUTLIER_HPO","SPLI_OUTLIER_RV","SPLI_OUTLIER_RV_CADD","SPLI_OUTLIER_RV_CADD_HPO")))

toplot=res_splicing[1:nrow(res_splicing),11:ncol(res_splicing)]
toplot=t(data.matrix(as.data.frame(toplot)))
percent=t(t(toplot[1:6,])/toplot[1,]*100)
percent=percent[-1,]
percent.m =melt(percent)


#to get percent of genes from intial pool of outlier
percent.m %>% filter(Var1=="SPLI_OUTLIER_RV_CADD_HPO") %>% mutate(average=mean(value, na.rm=T))

res_splicing_genes=mclapply(cases_withGen,get_genes_filters_splicing, mc.cores=10)
names(res_splicing_genes)=cases_withGen
indiv_id=data.frame(STAN_ID=cases_withGen)
indiv_id=indiv_id %>% left_join(metadata, by=c("STAN_ID"="sample_id")) %>% select(STAN_ID,institution_id)

to_print=cases_withGen[-c(7,8,17,18,27,28,37,38,47,48,57,58,67,68)]

indiv_id=data.frame(STAN_ID=to_print)
indiv_id=indiv_id %>% left_join(metadata, by=c("STAN_ID"="sample_id")) %>% select(STAN_ID,institution_id)

for (i in c(1:length(to_print))){
	print(i)
	for (j in c(2:length(names(res_splicing_genes[[to_print[i]]])))) {
		write.table(res_splicing_genes[[to_print[i]]][[j]], file=paste0(dir,"/analysis/outlier_analysis/splicing/2018_12_3_paper_revisions/",indiv_id[i,2],"_",names(res_splicing_genes[[to_print[i]]][j]),"_splicing_filtered_results.tsv"), sep="\t", quote=F, row.names=F)
	}
}

# Save data
save.image(file = paste0(dir,"/analysis/manuscript/figures_revision/Figure3.in.RData"))
save.image(file = "figures_revision/Figure3.in.RData")


