#!/bin/R

# This scripts performs analyses for influence of differet filters on the number of expression outlier gene candidates for Figure 2
#Libraries
library(ggplot2)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(sva)
library(cowplot)
library(purrr)
library(broom)
library(reshape2)
library(ggfortify)
library(gridExtra)
library(data.table)
library(annotables)
library(corrplot)
library(RColorBrewer)
library(qvalue) 
library(ggpubr)


## FUNCTIONS


rm(list=ls())


# Master directory
dir = Sys.getenv('RARE_DIS_DIR')
## FUNCTIONS

source(paste(dir,"/scripts/manuscript_analyses/Figures_source.R". sep="")


## MAIN
# Get metadata
metadata = paste(dir,"/data/metadata/2018_06_12_Rare_Disease_Metadata.tsv",sep="")
metadata = read_tsv(metadata) %>% filter(in_freeze=="yes") %>% filter(source_of_RNA=="Blood")

# get list of all RD samples
RD_samples=metadata$sample_id

# get list of samples with genetic data
sample_with_data=metadata %>% filter(variant_data=="exome" | variant_data=="genome")%>% select(sample_id) %>% pull

# read in exac data
exac = as.data.frame(read.table(paste(dir,"/data/annotations/EXAC/forweb_cleaned_exac_r03_march16_z_data_pLI.txt", sep=""), sep = '\t', stringsAsFactors = F, header = T))
exac=exac %>% inner_join(grch37, by=c("gene"="symbol")) %>% select(ensgene,syn_z, mis_z, lof_z, pLI) 

# read in affected status data
#affected_status_df=data.frame(sample=metadata$sample_id,status=metadata$affected_status)
affected_status_df=read.table(paste(dir,"/analysis/outlier_analysis/splicing/for_freeze/sample_affected_status_freeze_RD_PIVUS.tsv", sep=""), header=F)
colnames(affected_status_df)=c('sample', 'status')


# read in phenotypic data
# Read in annotation files linking HPO terms genes and phenotypes
gene_to_pheno=paste(dir,"/data/metadata/gene_to_phenotype.tsv", sep="")
gene_to_pheno = read_tsv(gene_to_pheno)
colnames(gene_to_pheno)=c("entrez", "symbol", "DiseaseId")

pheno_annot=paste(dir,"/data/metadata/phenotype_annotation.tsv", sep="")
pheno_annot=read_tsv(pheno_annot)
pheno_annot=pheno_annot %>% select(HPO_terms_ID,database_INDEX, number_unknown)

# Read in and process alternative HPO files (generated by parse_hpoterms.py)
parent_terms=paste(dir,"/data/metadata/parent_terms_hpo_2018_03_16.txt", sep="")
child_terms=paste(dir,"/data/metadata/child_terms_hpo_2018_03_16.txt", sep="")
altid_terms=paste(dir,"/data/metadata/altid_terms_hpo_2018_03_16.txt", sep="")

altid_terms=as.data.frame(read.table(altid_terms, header=F, sep="\t"))
colnames(altid_terms)=c("HPO_ID", "ALT_ID")
child_terms=as.data.frame(read.table(child_terms, header=F, sep="\t"))
colnames(child_terms)=c("HPO_ID", "ALT_ID")
parent_terms=as.data.frame(read.table(parent_terms, header=F, sep="\t"))
colnames(parent_terms)=c("HPO_ID", "ALT_ID")


# generates list of HPO alternative IDs from the input files
alt_hpo_ids_list=lapply(altid_terms$HPO_ID,get_hpo_alt_id, altid_terms=altid_terms)
names(alt_hpo_ids_list)=altid_terms$HPO_ID
child_hpo_ids_list=lapply(child_terms$HPO_ID,get_hpo_alt_id, altid_terms=child_terms)
names(child_hpo_ids_list)=child_terms$HPO_ID
parent_hpo_ids_list=lapply(parent_terms$HPO_ID,get_hpo_alt_id, altid_terms=parent_terms)
names(parent_hpo_ids_list)=parent_terms$HPO_ID

# For every ID associated with the disease, get list of alternative Ids, child ids and parent ids
# make a list for each case with HPO terms

cases=metadata %>% filter(affected_status=="Case") %>% select(sample_id) %>% pull

# Get list of HPO terms from metadata file
HPOs=lapply(cases,get_list_hpo, metadata)
names(HPOs)=cases
cases_withGen=metadata %>% filter(variant_data !="none", affected_status=="Case")%>% select(sample_id) %>% pull

# Generate extended list of HPO terms for each sample
HPO_extended=lapply(cases, process_sample_ids_hpo, HPO_list=HPOs)
names(HPO_extended)=cases


# expression outlier data
exp_outlier=paste(dir,"/analysis/outlier_analysis/expression_level/outliers_zscore_pair_spline_05apr18.txt",sep="")
exp_outlier=as.data.frame(read.table(exp_outlier, header=T))


# expression outlier with rare variant (window method)
RV_outlier_exp_10kb=paste(dir,"/analysis/outlier_analysis/expression_level/outliers_rare_var_combined_10kb.txt",sep="")
RV_outlier_exp_10kb=as.data.frame(read.table(RV_outlier_exp_10kb, sep="\t", header=F))
colnames(RV_outlier_exp_10kb)=c("sample_id", "gene", "expressionZScore", "variantPos", "gnomadMAF", "raw_cadd", "phred_cadd", "variant_gene_pos")

RV_outlier_exp_10kb=RV_outlier_exp_10kb %>% left_join(grch37, by=c("gene"="ensgene"), all.x=T)%>%select(sample_id,gene,expressionZScore,gnomadMAF,phred_cadd,variant_gene_pos,variantPos, start,end)

# get distance of variant from gene
RV_outlier_exp_10kb=RV_outlier_exp_10kb%>% mutate(distance=ifelse((variantPos<=start&variantPos<=end)|(variantPos>=start&variantPos>=end),start-variantPos,0))

# make a column with singleton included or expluded form the analysis (window method)
RV_outlier_exp_10kb =RV_outlier_exp_10kb %>% mutate(gnomAD_AF_withsgl = replace(gnomadMAF, gnomadMAF==".", 0))
RV_outlier_exp_10kb =RV_outlier_exp_10kb %>% mutate(gnomAD_AF_nosgl = replace(gnomadMAF, gnomadMAF==".", NA))


# format variant data so that only the max observed allele frequency is kept for variant where several frequencies are available (window method)
RV_outlier_exp_10kb$gnomAD_AF_filt_withsgl=sapply(RV_outlier_exp_10kb$gnomAD_AF_withsgl,process_variant)
RV_outlier_exp_10kb$gnomAD_AF_filt_nosgl=sapply(RV_outlier_exp_10kb$gnomAD_AF_nosgl,process_variant)


#filter for RV with MAF<=0.1% 
maf=0.001
	# Excluding singletons
RV_outlier_exp_filt_nosgl_10kb=RV_outlier_exp_10kb %>% filter(!is.na(gnomAD_AF_filt_nosgl)) %>% filter(gnomAD_AF_filt_nosgl <=maf)
	# Including singletons
RV_outlier_exp_filt_withsgl_10kb=RV_outlier_exp_10kb %>% filter(!is.na(gnomAD_AF_filt_withsgl)) %>% filter(gnomAD_AF_filt_withsgl <=maf)


# Read in RIVER scores
river_file=paste(dir,"/data/udn/river_score/UDN_RIVERscore_ByGene.txt", sep="")
river=as.data.frame(read.table(river_file, sep="\t", header=T))

	#get list of genes with RIVER score >=0.85 for each sample
gene_river_list=sapply(sample_with_data,get_genes_with_highriver,river)

# Read in ASE data
ase_file=paste(dir,"/data/ase/June13/final_merged/ensembleids.ase.filtered.txt", sep="")
ase=read_tsv(ase_file, col_names=TRUE)
colnames(ase)=c("sample_id", colnames(ase)[2:length(colnames(ase))])

	# transform ase data 
ase=ase %>% mutate(ensgene=str_extract(Gene, "ENSG[0-9]+")) %>% select(sample_id,Chr,POS, "RefRatio" , "TotalReads","ensgene")%>% filter (! duplicated(ensgene))%>% mutate(RefRatio_trans=ifelse(RefRatio<=0.5,1-RefRatio,RefRatio))%>% filter (! RefRatio_trans==1)
ase=ase%>% group_by(ensgene)%>% mutate(max_RefRatio_trans=max(RefRatio_trans,na.rm=T), median_RefRatio_trans=median(RefRatio_trans, na.rm=T))%>% ungroup

	# get list of genes with ASE for each sample
gene_ase_list=sapply(sample_with_data,get_genes_with_ase,ase)


## Influence of filtering on expression outlier results
zscore_threshold=2
distance_threshold=10000
cadd=10
pli=0.9


sample_exp_outlier_filter_withsgl_10kb_up.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=0, direction="under"),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=pli, direction="under"),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb_exp,outlier_file=exp_outlier,HPOs_list=HPO_extended, pLI_thre=0, direction="under"),
	"raw.ASE"=sapply(sample_with_data, get_outlier_ase_genes, threshold=zscore_threshold, outlier_df=exp_outlier, gene_ase_list=gene_ase_list,pLI_thre=0, direction="under"),
	"raw.river"=sapply(sample_with_data, get_outlier_river_genes, threshold=zscore_threshold, outlier_df=exp_outlier, gene_river_list=gene_river_list,pLI_thre=0, direction="under"),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=0, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=0, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.ASE"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb_ase,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_ase_list, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.RIVER"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb_ase,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_river_list, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD.pLI"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD.ASE"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb_ase,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_ase_list,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))


save("sample_exp_outlier_filter_withsgl_10kb_up.df", file=paste(dir,"/analysis/outlier_analysis/expression_level/expression_filters_genes.RData", sep=""))


# make a plot with proportions of candidates 

sample_exp_outlier_filter_withsgl_10kb_up.df= sample_exp_outlier_filter_withsgl_10kb_up.df %>% filter(raw!=0)
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.Lof/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_HPO=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.HPO/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RIVER=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.river/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV=sample_exp_outlier_filter_withsgl_10kb_up.df$RV/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.LoF.intolerant/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_CADD=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.CADD/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.RIVER=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.RIVER/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD.pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD.pLI/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD.ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw

sample_exp_outlier_filter_withsgl_10kb_up.df=sample_exp_outlier_filter_withsgl_10kb_up.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_RIVER, prop_ASE, prop_HPO,prop_RV,prop_RV.pLI,prop_CADD,prop_RV.RIVER, prop_RV.HPO, prop_RV.ASE,prop_RV.HPO.CADD,prop_RV.HPO.CADD.pLI,prop_RV.HPO.CADD.ASE)

#melt DF
sample_exp_outlier_filter_withsgl_10kb_up.df.m=melt(sample_exp_outlier_filter_withsgl_10kb_up.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_exp_outlier_filter_withsgl_10kb_up.df.m.cases=sample_exp_outlier_filter_withsgl_10kb_up.df.m %>% filter(affected_status=="Case")


# Save data
save.image(file = paste(dir,"/data/Figure2D.in.RData",sep=""))


