#!/bin/R
# Feb 2018
# LF
# Figure panel on splicing analysis

# A: overview of the model (included separately)
# B: Number of outliers at different Zscores
# C: Prop of outliers left when filtering for rare variants nearby
# D: Number of RV before and after filtering for splciing outliers




#Libraries
library(ggplot2)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(sva)
library(cowplot)
library(purrr)
library(broom)
library(reshape2)
library(ggfortify)
library(gridExtra)
library(data.table)
library(annotables)
library(corrplot)
library(RColorBrewer)
library(qvalue) 
library(ggpubr)
library(missMDA)

fsize=15
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize),
	axis.text.y= element_text(size=fsize), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.1),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=9),
	panel.border=element_blank()) 

rm(list=ls())

source('/users/lfresard/repos/rare_disease/scripts/manuscript_analyses/Figures_source.R')


# Main

#--- MAIN
# Get metadata
metadata = "/srv/scratch/restricted/rare_diseases/data/metadata/2018_06_12_Rare_Disease_Metadata.tsv"
metadata = read_tsv(metadata) %>% filter(in_freeze=="yes") %>% filter(source_of_RNA=="Blood")

# Get metadata info for pivus and dgn
meta_pivus="/srv/scratch/restricted/rare_diseases/data/metadata/PIVUS_RNASequencingInfo.csv"
meta_pivus=read_csv(meta_pivus) %>% filter(Age=="70") %>% filter(RunOK=="Yes")
meta_pivus2="/srv/scratch/restricted/rare_diseases/data/metadata/Pivus_expCovariatesAll_unscaled.txt"
meta_pivus2=read_tsv(meta_pivus2)%>% filter(Age==0) #%>% filter(RunOK=="Yes")


#samples= metadata %>% select(sample_id) %>% distinct %>% pull

# select samples for which we have genetic data
sample_with_data=metadata %>% filter(variant_data=="exome" | variant_data=="genome")%>% select(sample_id) %>% pull
sample_with_data=sample_with_data[!sample_with_data=="RD008"]


# Get annotated junctions
annot_junc="/srv/scratch/restricted/rare_diseases/data/splicing/juncfiles/gencodev19_intronsstartplus1_proteincoding.genenames_uniqjunc.tsv"
annot_junc=read_tsv(annot_junc, col_names=FALSE)
colnames(annot_junc)=c("chr", "junc_start", "junc_end", "gene")
annot_junctions=paste(annot_junc$chr, annot_junc$junc_start, annot_junc$junc_end, annot_junc$gene, sep="_")


## affected status data
affected_status_df=read.table('/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/ifrun9/sample_affected_status_freeze_RD_PIVUS.tsv', header=F)
colnames(affected_status_df)=c('sample', 'status')

affected_status_df$cohort=c(rep("RD",87), rep("PIVUS", 68))


samples=affected_status_df$sample
RD_samples=metadata$sample_id


# read in exac data
exac = as.data.frame(read.table('/users/lfresard/NumberOfIndividuals_Impact/enrichment_disease-conservation_database/annotations/EXAC/forweb_cleaned_exac_r03_march16_z_data_pLI.txt', sep = '\t', stringsAsFactors = F, header = T))
exac=exac %>% inner_join(grch37, by=c("gene"="symbol")) %>% select(ensgene,syn_z, mis_z, lof_z, pLI) 



# Read in annotation files linking HPO terms genes and phenotypes
gene_to_pheno="/srv/scratch/restricted/rare_diseases/data/metadata/gene_to_phenotype.tsv"
gene_to_pheno = read_tsv(gene_to_pheno)
colnames(gene_to_pheno)=c("entrez", "symbol", "DiseaseId")

pheno_annot="/srv/scratch/restricted/rare_diseases/data/metadata/phenotype_annotation.tsv"
pheno_annot=read_tsv(pheno_annot)
pheno_annot=pheno_annot %>% select(HPO_terms_ID,database_INDEX, number_unknown)

# Read in and process alternative HPO files (generated by parse_hpoterms.py)
parent_terms="/srv/scratch/restricted/rare_diseases/data/metadata/parent_terms_hpo_2018_03_16.txt"
child_terms="/srv/scratch/restricted/rare_diseases/data/metadata/child_terms_hpo_2018_03_16.txt"
altid_terms="/srv/scratch/restricted/rare_diseases/data/metadata/altid_terms_hpo_2018_03_16.txt"

altid_terms=as.data.frame(read.table(altid_terms, header=F, sep="\t"))
colnames(altid_terms)=c("HPO_ID", "ALT_ID")
child_terms=as.data.frame(read.table(child_terms, header=F, sep="\t"))
colnames(child_terms)=c("HPO_ID", "ALT_ID")
parent_terms=as.data.frame(read.table(parent_terms, header=F, sep="\t"))
colnames(parent_terms)=c("HPO_ID", "ALT_ID")


# generates list of HPO alternative IDs from the input files
alt_hpo_ids_list=lapply(altid_terms$HPO_ID,get_hpo_alt_id, altid_terms=altid_terms)
names(alt_hpo_ids_list)=altid_terms$HPO_ID
child_hpo_ids_list=lapply(child_terms$HPO_ID,get_hpo_alt_id, altid_terms=child_terms)
names(child_hpo_ids_list)=child_terms$HPO_ID
parent_hpo_ids_list=lapply(parent_terms$HPO_ID,get_hpo_alt_id, altid_terms=parent_terms)
names(parent_hpo_ids_list)=parent_terms$HPO_ID

# For every ID associated with the disease, get list of alternative Ids, child ids and parent ids
# make a list for each case with HPO terms

cases=metadata %>% filter(affected_status=="Case") %>% select(sample_id) %>% pull

# Get list of HPO terms from metadata file
HPOs=lapply(cases,get_list_hpo, metadata)
names(HPOs)=cases
cases_withGen=metadata %>% filter(variant_data !="none", affected_status=="Case")%>% select(sample_id) %>% pull

# Generate extended list of HPO terms for each sample
HPO_extended=lapply(cases, process_sample_ids_hpo, HPO_list=HPOs)
names(HPO_extended)=cases


HPO_extended <- Map(cbind, HPO_extended, sample = names(HPO_extended))
HPO_extended.df=as.data.frame(do.call("rbind", HPO_extended))
colnames(HPO_extended.df)=c("HPO_term", "sample")
HPO_extended.df$sample=as.character(HPO_extended.df$sample)


# attach gene to HPO ids for each samples
HPO_extended.df=HPO_extended.df %>% left_join(pheno_annot, by=c("HPO_term"="HPO_terms_ID"), all.left=T) %>% select(HPO_term,sample,database_INDEX) %>% left_join(gene_to_pheno, by=c("database_INDEX"="DiseaseId"), all.left=T) %>% select(HPO_term,sample,database_INDEX,symbol) %>% left_join(grch37, by="symbol", all.left=T)  %>% select(HPO_term,sample, ensgene) %>% unique %>% filter(!is.na(ensgene))

## get ratio data
#ratio_file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/RD_freeze_junc_ratios.txt"
ratio_file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/ifrun9/RD_PIVUS_freeze_junc_ratios.txt"
ratios=as.data.frame(read.table(ratio_file,  header=T, sep="\t"))

all_samples=c(metadata$sample_id, meta_pivus$RNAseq_ID)


ratios=ratios %>% select(-Cluster) %>% distinct
# make junction names out of coordinates
junctions=paste(ratios$chr,ratios$junction_start, ratios$junction_end, ratios$gene, sep="_")
#samples_batch8=metadata%>% filter(batch==8)%>% select(sample_id) %>% pull


ratio.df=data.frame( ratios[,all_samples])
rownames(ratio.df)=junctions
ratio.df[is.na(ratio.df)]<-NA


# remove junctiosn for which we have less than 30 samples with any value
ratio.df=ratio.df[colSums(!is.na(t(ratio.df)))>30,]

na.data=is.na(ratio.df)

#nb = estim_ncpPCA(t(ratio.df))
res.comp = imputePCA(t(ratio.df),ncp=10)
pca=prcomp(res.comp$completeObs)


sum_pca=t(summary(pca)$importance)
colnames(sum_pca)=c("sd", "prop_var", "cumul_var")
pcs_number=as.data.frame(sum_pca) %>% filter(cumul_var <=0.95) %>% nrow
pcs_selec=pca$x[,1:pcs_number]
mod = model.matrix(~1, data=as.data.frame(res.comp$completeObs))
#mod = model.matrix(~1, data=as.data.frame(t(ratio.df)))
modsv <- cbind(mod, pcs_selec)
fitsv <- lm.fit(modsv, res.comp$completeObs)


save(res.comp,fitsv,pcs_selec, file= "PCA_splicing.RData")


load("PCA_splicing.RData")
junction_zscore_matrix=as.matrix(fitsv$residuals)
junction_zscore_matrix.t=t(junction_zscore_matrix)
junction_zscore_matrix.t[na.data]<-NA
junction_zscore_matrix=t(junction_zscore_matrix.t)

## Scale and center
junction_zscore_scale =scale(junction_zscore_matrix, center=TRUE, scale=TRUE)
junction_zscore_scale.t=t(junction_zscore_scale)
junction_zscore_scale.t.df=as.data.frame(junction_zscore_scale.t)

junction_zscore_scale.t.df$annotation_status=sapply(rownames(junction_zscore_scale.t.df),is_annotated, junc_ref_vect=annot_junctions)
junction_zscore_scale.t.df$chr=unlist(strsplit(rownames(junction_zscore_scale.t.df), "_"))[c(TRUE,FALSE,FALSE,FALSE)]
junction_zscore_scale.t.df$junc_start=unlist(strsplit(rownames(junction_zscore_scale.t.df), "_"))[c(FALSE,TRUE,FALSE,FALSE)]
junction_zscore_scale.t.df$junc_end=unlist(strsplit(rownames(junction_zscore_scale.t.df), "_"))[c(FALSE,FALSE,TRUE,FALSE)]
junction_zscore_scale.t.df$gene=unlist(strsplit(rownames(junction_zscore_scale.t.df), "_"))[c(FALSE,FALSE,FALSE, TRUE)]


junction_zscore_scale.t.df.m=melt(junction_zscore_scale.t.df, id.var=c("chr", "junc_start", "junc_end","gene", "annotation_status"))
junction_zscore_scale.t.df.m$absZ=abs(junction_zscore_scale.t.df.m$value)
zscores_blood=junction_zscore_scale.t.df.m %>% left_join(affected_status_df, by=c("variable"="sample")) %>% select(chr ,junc_start  ,junc_end, gene ,annotation_status ,variable, value, absZ, status)




# correct ratio data 
	#zscores_blood=get_corrected_data(affected_status_df,ratio.df)
# integrate previous analysis in function
# filterzscore_blood for samples with genetic data
zscores_blood_gendata=zscores_blood %>% filter(variable %in% sample_with_data)

# write down results to filter for rare variants nearby
write.table(zscores_blood_gendata,'RD_freeze_junc_outliers_PCAratio_blood_nf_PIVUS_sampwithgen.txt', col.names=T, row.names=F, quote=F, sep="\t", append=F)
write.table(zscores_blood,'RD_freeze_junc_outliers_PCAratio_blood_nf_PIVUS_withPIVsamples_withmisdata.txt', col.names=T, row.names=F, quote=F, sep="\t", append=F)



# remove suffix from gene names

zscores_blood =as.data.frame(read.table('RD_freeze_junc_outliers_PCAratio_blood_nf_PIVUS_withPIVsamples_withmisdata.txt', sep="\t", header=T))

zscores_blood =zscores_blood %>% mutate(gene=unlist(strsplit(as.character(gene),"[.]"))[c(TRUE,FALSE)]) 
genes=zscores_blood %>% select(gene) %>% distinct %>% pull

max_z_out=zscores_blood %>% filter(absZ>=2.5)%>% group_by(gene,variable) %>% mutate(maxZ=max(absZ, na.rm=T)) %>% as.data.frame %>% select(gene,variable,status,maxZ) %>% unique
max_z_out$maxZ[max_z_out$maxZ==-Inf] <-NA
write.table(max_z_out,'splicing_outlier_maxZ.txt', col.names=T, row.names=F, quote=F, sep="\t", append=F)
max_z_out_for_amelie=max_z_out%>% filter(status=="Case") %>% left_join(metadata[,c("sample_id","HPO_terms_ID")], by=c("variable"="sample_id")) %>% group_by(variable, HPO_terms_ID) %>% summarise(genes = toString(gene)) %>% ungroup
write.table(max_z_out_for_amelie,'splicing_outlier_maxZ_for_amelie.txt', col.names=T, row.names=F, quote=F, sep="\t", append=F)



meta_samples= metadata %>% select(sample_id, institution_id, institution)
grch37_simple=grch37 %>% select(ensgene, symbol) %>% distinct(ensgene, symbol)
outliers_for_UDN=zscores_blood %>% left_join(meta_samples, by=c("variable"="sample_id"))%>%left_join(grch37_simple, by=c("gene"="ensgene")) %>%filter(institution=="UDN", status=="Case")%>% arrange(variable,-absZ) 
write.table(outliers_for_UDN,'splicing_outliers_for_UDN.txt', col.names=T, row.names=F, quote=F, sep="\t", append=F)



## Analyze results filtered by RV

#RV_outlier_file='./RV/RD_freeze_junc_outliers_PCAratio_blood_nf_PIVUS_withPIVsamples_withmisdata_sorted_RV.txt'
RV_outlier_file='./RV/RD_freeze_junc_outliers_PCAratio_blood_nf_PIVUS_withPIVsamples_withmisdata_sorted_RV_window.txt'
RV_outlier=as.data.frame(read.table(RV_outlier_file, sep="\t", header=F))
#colnames(RV_outlier)=c(colnames(zscores_blood), "chr_var", "start_var", "end_var", "ref_var", "alt_var", "gnomAD_AF", "raw_cadd", "phred_cadd", "chr_gene", "start_gene", "end_gene", "gene_name", "distance_variant_junction")
colnames(RV_outlier)=c(colnames(zscores_blood), "chr_var", "start_var", "end_var", "ref_var", "alt_var", "gnomAD_AF", "raw_cadd", "phred_cadd", "chr_gene", "start_gene", "end_gene", "gene_name")

# make a column with singleton included or expluded form the analysis
RV_outlier =RV_outlier %>% mutate(gnomAD_AF_withsgl = replace(gnomAD_AF, gnomAD_AF==".", 0))
RV_outlier =RV_outlier %>% mutate(gnomAD_AF_nosgl = replace(gnomAD_AF, gnomAD_AF==".", NA))


# format variant data so that only the max observed allele frequency is kept for variant were several frequencies are available
RV_outlier$gnomAD_AF_filt_withsgl=sapply(RV_outlier$gnomAD_AF_withsgl,process_variant)
RV_outlier$gnomAD_AF_filt_nosgl=sapply(RV_outlier$gnomAD_AF_nosgl,process_variant)

#filter for RV with MAF<=0.1%
maf=0.001
	# Excluding singletons
RV_outlier_filt_nosgl=RV_outlier %>% filter(!is.na(gnomAD_AF_filt_nosgl)) %>% filter(gnomAD_AF_filt_nosgl <=maf)
RV_outlier_filt_nosgl =RV_outlier_filt_nosgl %>% mutate(gene_name=str_extract(gene_name, "ENSG[0-9]+")) 
	# Including singletons
RV_outlier_filt_withsgl=RV_outlier %>% filter(!is.na(gnomAD_AF_filt_withsgl)) %>% filter(gnomAD_AF_filt_withsgl <=maf)
RV_outlier_filt_withsgl =RV_outlier_filt_withsgl %>% mutate(gene_name=str_extract(gene_name, "ENSG[0-9]+"))

RV_outlier_filt_withsgl_indels=RV_outlier_filt_withsgl %>% filter(nchar(as.character(alt_var))>1 | nchar(as.character(ref_var))>1)
RV_outlier_filt_withsgl_snps=RV_outlier_filt_withsgl %>% filter(nchar(as.character(alt_var))==1 & nchar(as.character(ref_var))==1)
# Make dataframe containing the number of genes with at least 1 outlier at different Zscore thresholds

sample_outlier.df=data.frame(sample_id=all_samples,
	"Z2"=sapply(all_samples, get_outlier_genes, threshold=2, outlier_df.m=zscores_blood, 0),
	"Z3"=sapply(all_samples, get_outlier_genes, threshold=3, outlier_df.m=zscores_blood, 0),
	"Z4"=sapply(all_samples, get_outlier_genes, threshold=4, outlier_df.m=zscores_blood, 0),
	"Z5"=sapply(all_samples, get_outlier_genes, threshold=5, outlier_df.m=zscores_blood, 0),
	"Z6"=sapply(all_samples, get_outlier_genes, threshold=6, outlier_df.m=zscores_blood, 0),
	#"Z7"=sapply(samples, get_outlier_genes, threshold=7, outlier_df.m=zscores_blood),
	#institution=sapply(samples,get_institution, metadata=affected_status_df),
	affected_status=sapply(all_samples,get_affected_status, affected_status_df=affected_status_df))
	#technology=rep("no filter", length(samples)))

# Add cohort information to data frame
sample_outlier.df=sample_outlier.df %>% left_join(affected_status_df, by=c("sample_id"="sample")) %>% select(-status) 

# transform data for plotting
sample_outlier.df.m=melt(sample_outlier.df,id.vars=c("sample_id","affected_status", "cohort"))
sample_outlier.df.m$variable=c(rep(2,152),rep(3,152),rep(4,152), rep(5,152), rep(6,152))

# plot the number of genes with at least one outlier at different zscore thresholds
# for poster
fsize=20
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize),
	axis.text.y= element_text(size=fsize), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.5),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=fsize),
	panel.border=element_blank()) 


zscore_thres_case_control_plot=ggplot(sample_outlier.df.m, aes(x=as.factor(variable), y=value+1, fill=affected_status))+
	geom_boxplot()+
	#stat_summary(fun.data=data_summary, geom="pointrange", color="black", size=0.8,  position = position_dodge(0.9))+
	scale_fill_brewer(palette="Paired",name="Affected status")+
	scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x)))+
	labs(x="Zscore threshold", y="Number of genes with\nat least one outlier")+RD_theme +theme(legend.position=c(0.8,0.5))+ stat_compare_means(aes(group = affected_status), label = "p.signif")
zscore_thres_case_control_plot

zscore_thres_cohort=ggplot(sample_outlier.df.m, aes(x=as.factor(variable), y=value+1, fill=cohort))+
	geom_boxplot()+
	#stat_summary(fun.data=data_summary, geom="pointrange", color="black", size=0.8,  position = position_dodge(0.9))+
	scale_fill_brewer(palette="Reds",name="Cohort")+
	scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x)))+
	labs(x="Z-score threshold", y="Number of genes with\nat least one outlier")+RD_theme +theme(legend.position=c(0.8,0.9))+ stat_compare_means(aes(group = cohort), label = "p.signif")

ggsave('fig3b_PIVUS.pdf', fig_3b_zscore_thres, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
ggsave('sup_zscore_thres_cohorteffect.pdf', zscore_thres_cohort, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
ggsave('sup_zscore_thres_case_control.pdf', zscore_thres_case_control_plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)


# RD theme wth fsize=20

fig_3b_zscore_thres_poster=ggplot(sample_outlier.df.m, aes(x=as.factor(variable), y=value+1))+
	geom_boxplot(color="black", fill="grey")+#fill="#028482")+
	#stat_summary(fun.y=mean, geom="point", size=2, color="#028482")+ 
	#stat_summary(fun.data=data_summary, geom="pointrange", color="black", size=1)+
	#scale_fill_brewer(palette="Paired",name="Affected status")+
	scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x)))+
	labs(x="Z-score threshold", y="Number of genes with\nat least one outlier")+RD_theme #+theme(legend.position="bottom")
fig_3b_zscore_thres_poster


ggsave('fig3b_PIVUS_poster.pdf', fig_3b_zscore_thres_poster, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)



## combined_plot
sup_zscore_thres_plot=ggdraw()+draw_plot(zscore_thres_case_control_plot, 0,1/2,1,1/2)+
	draw_plot(zscore_thres_cohort, 0,0,1,1/2)+
	draw_plot_label(c('A', 'B'), c(0,0), c(1,1/2), size = 20)

sup_zscore_thres_plot

pdf('/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/sup_zscore_thres_plot.pdf', w=8, h=10)
sup_zscore_thres_plot
dev.off()


# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# Figure 3C
# number of RV
RV_junc_dir='/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/RV/'
#RV_junc_samples=list.files(pattern='_junction_RV.txt', path=RV_junc_dir)
RV_junc_samples=list.files(pattern='_junction_RV_window.txt', path=RV_junc_dir)


bed_dir='/srv/scratch/restricted/rare_diseases/data/bed/for_freeze_exac_filt/'

# count the number of RV nearby annotated junctions for each sample
	#with singletons
#RV_junction_number=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0,20, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0.001,20, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0.01,20, "yes"))
RV_junction_number=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.01, "yes"))
RV_junction_number=RV_junction_number %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_junction_number=RV_junction_number%>% mutate(filter="RV near junction")%>% mutate(splicing_data="no")%>% mutate(singleton="yes")

	# without singletons
#RV_junction_number_nosgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0,20, "no"), RV_number_MAF0.1=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0.001,20, "no"),RV_number_MAF1=sapply(sample_with_data, get_number_junction_RV, RV_junc_dir, 0.01,20, "no"))
RV_junction_number_nosgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0, "no"), RV_number_MAF0.1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.001, "no"),RV_number_MAF1=sapply(sample_with_data, get_number_junction_RV_window, RV_junc_dir, 0.01, "no"))
RV_junction_number_nosgl=RV_junction_number_nosgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_junction_number_nosgl=RV_junction_number_nosgl%>% mutate(filter="RV near junction")%>% mutate(splicing_data="no")%>% mutate(singleton="no")


# count the number of RV  for each sample

RV_number_withsgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV, bed_dir, 0, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV, bed_dir, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_RV, bed_dir, 0.01, "yes"))
RV_number_withsgl=RV_number_withsgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_number_withsgl=RV_number_withsgl%>% mutate(filter="Rare variant")%>% mutate(splicing_data="no")%>% mutate(singleton="yes")



RV_number_nosgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV, bed_dir, 0, "no"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV, bed_dir, 0.001, "no"),RV_number_MAF1=sapply(sample_with_data, get_number_RV, bed_dir, 0.01, "no"))
RV_number_nosgl=RV_number_nosgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)#%>% rename(variant_data=variant_data.x, affected_status=affected_status.x)
RV_number_nosgl=RV_number_nosgl %>% mutate(filter="Rare variant")%>% mutate(splicing_data="no")%>% mutate(singleton="no")

# data with splicing outliers
#RV_number_out_withsgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0, 20, "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0.001,20, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0.01,20, "yes"))
RV_number_out_withsgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0,  "yes"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.001, "yes"),RV_number_MAF1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.01, "yes"))
RV_number_out_withsgl=RV_number_out_withsgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)
RV_number_out_withsgl=RV_number_out_withsgl %>% mutate(filter="splicing outlier")%>% mutate(splicing_data="yes")%>% mutate(singleton="yes")

#RV_number_out_nosgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0, 20,"no"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0.001,20, "no"),RV_number_MAF1=sapply(sample_with_data, get_number_RV_outlier,RV_outlier,3, 0.01,20, "no"))
RV_number_out_nosgl=data.frame(sample=sample_with_data, RV_number_MAF0=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0,"no"), RV_number_MAF0.1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.001, "no"),RV_number_MAF1=sapply(sample_with_data, get_number_RV_outlier_window,RV_outlier,3, 0.01, "no"))
RV_number_out_nosgl=RV_number_out_nosgl %>% left_join(metadata,by=c("sample"="sample_id")) %>% select(sample, RV_number_MAF0,RV_number_MAF0.1,RV_number_MAF1,variant_data, affected_status)
RV_number_out_nosgl=RV_number_out_nosgl %>% mutate(filter="splicing outlier")%>% mutate(splicing_data="yes")%>% mutate(singleton="no")


# combine info of RV and splicing outlier
	RV_outliers=rbind(RV_junction_number,RV_junction_number_nosgl,RV_number_nosgl,RV_number_withsgl,RV_number_out_nosgl,RV_number_out_withsgl)
#RV_outliers=rbind(RV_junction_number,RV_number_withsgl,RV_number_out_withsgl)

# melt DF for plot
RV_outliers.m=melt(RV_outliers,id.var=c("sample", "variant_data", "affected_status", "filter", "splicing_data", "singleton"))

RV_outliers.m_nosgl=RV_outliers.m %>% filter(singleton=="no", variable!="RV_number_MAF0")
RV_outliers.m_withsgl=RV_outliers.m %>% filter(singleton=="yes", variable!="RV_number_MAF0")

col_splic_out=c( "#87907D"	,"#AAB6A2", 	"#666666")
RV_out_plot_nosgl=ggplot(RV_outliers.m_nosgl, aes(x=variable, y=value+1, fill=filter)) + 
	geom_boxplot()+ 
	scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x)))+
	scale_fill_manual(values=col_splic_out, name="Data filtered for\nsplicing outliers")+
	scale_x_discrete(breaks=c("RV_number_MAF0", "RV_number_MAF0.1", "RV_number_MAF1"), labels=c("0", "0.1", "1"))+
	labs(x="MAF (%)", y="Number of rare variants")+facet_grid(  . ~variant_data , scales="free_y")+RD_theme +
	theme(legend.position=c(0.4,0.9))


RV_out_plot_nosgl
ggsave('sup_RV_outlier_nosgltons.pdf', RV_out_plot_nosgl, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=10, height=10)


# with singletons
RV_out_plot_withsgl=ggplot(RV_outliers.m_withsgl, aes(x=variable, y=value+1, fill=filter)) + 
geom_boxplot()+ 
scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),labels = scales::trans_format("log10", scales::math_format(10^.x)))+
scale_fill_manual(values=col_splic_out, name="Data filtered for\nsplicing outliers")+
scale_x_discrete(breaks=c("RV_number_MAF0", "RV_number_MAF0.1", "RV_number_MAF1"), labels=c("0", "0.1", "1"))+
labs(x="MAF (%)", y="Number of rare variants")+facet_grid(  . ~variant_data , scales="free_y")+RD_theme +theme(legend.position=c(0.4,0.9))

RV_out_plot_withsgl
ggsave('Fig3C_RV_out_plot.pdf', RV_out_plot_withsgl, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=10, height=10)

#ggsave('Sup_RV_out_plot_withsgl.pdf', RV_out_plot_withsgl, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)


# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# Figure 3 D
## Influence of filtering on splicing outlier results
zscore_threshold=2
distance_threshold=20
cadd=10
pli=0.9

sample_outlier_filter_nosgl.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=0),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=pli),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb,outlier_file=zscores_blood,HPOs_list=HPO_extended, pLI_thre=0),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_nosgl, CADD=0, pLI_thre=0),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_nosgl,CADD=0, pLI_thre=pli),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_nosgl, CADD=cadd, pLI_thre=0),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_nosgl, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_nosgl, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0),
	"combined"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_nosgl,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))


sample_outlier_filter_withsgl.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=0),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=pli),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb,outlier_file=zscores_blood,HPOs_list=HPO_extended, pLI_thre=0),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl, CADD=0, pLI_thre=0),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl, CADD=0, pLI_thre=pli),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl,CADD=cadd, pLI_thre=0),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0),
	"combined"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))

save("sample_outlier_filter_withsgl.df", file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/splicing_filters_genes.RData")


# make a plot with proportions of candidates 

sample_outlier_filter_withsgl.df= sample_outlier_filter_withsgl.df %>% filter(raw!=0)
sample_outlier_filter_withsgl.df$prop_pLI=sample_outlier_filter_withsgl.df$raw.Lof/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_HPO=sample_outlier_filter_withsgl.df$raw.HPO/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RV=sample_outlier_filter_withsgl.df$RV/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVpLI=sample_outlier_filter_withsgl.df$RV.LoF.intolerant/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_CADD=sample_outlier_filter_withsgl.df$RV.CADD/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVHPO=sample_outlier_filter_withsgl.df$RV.HPO/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_RVHPOCADD=sample_outlier_filter_withsgl.df$RV.HPO.CADD/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df$prop_combined=sample_outlier_filter_withsgl.df$combined/sample_outlier_filter_withsgl.df$raw
sample_outlier_filter_withsgl.df=sample_outlier_filter_withsgl.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_HPO,prop_RV,prop_RVpLI,prop_CADD, prop_RVHPO,prop_RVHPOCADD,prop_combined)

#melt DF
sample_outlier_filter_withsgl.df.m=melt(sample_outlier_filter_withsgl.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_outlier_filter_withsgl.df.m.cases=sample_outlier_filter_withsgl.df.m %>% filter(affected_status=="Case")

# plot results
filter_withsglt.case.plot=ggplot(sample_outlier_filter_withsgl.df.m.cases, aes(x=variable, y=value, fill=variable) )+ 
	geom_boxplot(color="black", notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 8), breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c(expression("pLI " >="0.9"), "HPO match", expression("Rare variant within 20bp","1 + 3","3 + CADD score  " >= "10"), "2 + 3", "2 + 5", "1 + 2 + 3 + 5"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c("1", "2", "3","4","5", "6", "7", "8"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56))))+theme(legend.position = c(0.8,0.8))
filter_withsglt.case.plot
ggsave('fig_3d_prop_candidates_filters.pdf', filter_withsglt.case.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
ggsave('fig_3d_prop_candidates_filters_poster.pdf', filter_withsglt.case.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
fsize=20



filter_withsglt.casecontrol.plot=ggplot(sample_outlier_filter_withsgl.df.m, aes(x=variable, y=value, fill=variable, color=affected_status) )+ 
	geom_boxplot( notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_colour_brewer(palette="Paired",name="Affected status")+
	scale_fill_manual(values=rep("lightgrey", 8), breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c(expression("pLI " >="0.9"), "HPO match", expression("Rare variant within 20bp","1 + 3","3 + CADD score  " >= "10"), "2 + 3", "2 + 5", "1 + 7"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c("1", "2", "3","4","5", "6", "7", "8"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56))), colour=guide_legend(override.aes = list(size = 4,shape =c(1,1))))+theme(legend.position = c(0.8,0.8))
#filter_withsglt.case.plot
ggsave('sup_prop_candidates_filters_cases_controls.pdf', filter_withsglt.casecontrol.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

filter_withsglt.technology.plot=ggplot(sample_outlier_filter_withsgl.df.m, aes(x=variable, y=value, fill=variable, color=technology) )+ 
	geom_boxplot( notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_color_manual(values=c("#003366", "#CCCC99"))+ 
	scale_fill_manual(values=rep("lightgrey", 8), breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c(expression("pLI " >="0.9"), "HPO match", expression("Rare variant within 20bp","1 + 3","3 + CADD score  " >= "10"), "2 + 3", "2 + 5", "1 + 7"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c("1", "2", "3","4","5", "6", "7", "8"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56))), colour=guide_legend(override.aes = list(size = 4,shape =c(1,1))))+theme(legend.position = c(0.8,0.8))
#filter_withsglt.case.plot
ggsave('sup_prop_candidates_filters_technology.pdf', filter_withsglt.technology.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

save(filter_withsglt.technology.plot,filter_withsglt.case.plot, filter_withsglt.casecontrol.plot,file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/spli_filters.Rdata")


## InDELS
sample_outlier_filter_withsgl_indels.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=0),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes, threshold=zscore_threshold, outlier_df.m=zscores_blood, pLI_thre=pli),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb,outlier_file=zscores_blood,HPOs_list=HPO_extended, pLI_thre=0),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl_snps, CADD=0, pLI_thre=0),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl_indels, CADD=0, pLI_thre=pli),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_window, threshold=zscore_threshold, outlier_df.m=RV_outlier_filt_withsgl_indels,CADD=cadd, pLI_thre=0),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl_indels, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl_indels, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0),
	"combined"=sapply(sample_with_data, get_hpo_match_nb_RV_window,  outlier_RV=RV_outlier_filt_withsgl_indels,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))



sample_outlier_filter_withsgl_indels.df= sample_outlier_filter_withsgl_indels.df %>% filter(raw!=0)
sample_outlier_filter_withsgl_indels.df$prop_pLI=sample_outlier_filter_withsgl_indels.df$raw.Lof/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_HPO=sample_outlier_filter_withsgl_indels.df$raw.HPO/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_RV=sample_outlier_filter_withsgl_indels.df$RV/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_RVpLI=sample_outlier_filter_withsgl_indels.df$RV.LoF.intolerant/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_CADD=sample_outlier_filter_withsgl_indels.df$RV.CADD/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_RVHPO=sample_outlier_filter_withsgl_indels.df$RV.HPO/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_RVHPOCADD=sample_outlier_filter_withsgl_indels.df$RV.HPO.CADD/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df$prop_combined=sample_outlier_filter_withsgl_indels.df$combined/sample_outlier_filter_withsgl_indels.df$raw
sample_outlier_filter_withsgl_indels.df=sample_outlier_filter_withsgl_indels.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_HPO,prop_RV,prop_RVpLI,prop_CADD, prop_RVHPO,prop_RVHPOCADD,prop_combined)

#melt DF
sample_outlier_filter_withsgl_indels.df.m=melt(sample_outlier_filter_withsgl_indels.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_outlier_filter_withsgl_indels.df.m.cases=sample_outlier_filter_withsgl_indels.df.m %>% filter(affected_status=="Case")

# plot results
filter_withsglt_indels.case.plot=ggplot(sample_outlier_filter_withsgl_indels.df.m.cases, aes(x=variable, y=value, fill=variable) )+ 
	geom_boxplot(color="black", notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 8), breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c(expression("pLI " >="0.9"), "HPO match", expression("Rare variant within 20bp","1 + 3","3 + CADD score  " >= "10"), "2 + 3", "2+ 5", "1 + 7"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_HPO","prop_RV", "prop_RVpLI", "prop_CADD","prop_RVHPO", "prop_RVHPOCADD","prop_combined"),
		labels=c("1", "2", "3","4","5", "6", "7", "8"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56))))+theme(legend.position = c(0.8,0.8))
#filter_withsglt.case.plot
ggsave('sup_prop_candidates_filters_indels.pdf', filter_withsglt_indels.case.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

sample_outlier_filter_nosgl.df= sample_outlier_filter_nosgl.df %>% filter(raw!=0)
sample_outlier_filter_nosgl.df$prop_pLI=sample_outlier_filter_nosgl.df$raw.Lof/sample_outlier_filter_nosgl.df$raw
sample_outlier_filter_nosgl.df$prop_RV=sample_outlier_filter_nosgl.df$RV/sample_outlier_filter_nosgl.df$raw
sample_outlier_filter_nosgl.df$prop_RVpLI=sample_outlier_filter_nosgl.df$RV.LoF.intolerant/sample_outlier_filter_nosgl.df$raw
sample_outlier_filter_nosgl.df$prop_CADD=sample_outlier_filter_nosgl.df$RV.CADD/sample_outlier_filter_nosgl.df$raw
sample_outlier_filter_nosgl.df$prop_combined=sample_outlier_filter_nosgl.df$combined/sample_outlier_filter_nosgl.df$raw



sample_outlier_filter_nosgl.df=sample_outlier_filter_nosgl.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_RV,prop_RVpLI,prop_CADD, prop_combined)
#sample_outlier_filter.df2=sample_outlier_filter.df %>% select(sample_id,institution,affected_status, technology,prop_pLI, prop_RV,prop_RVpLI, prop_CADD) #%>% mutate(prop_init=rep(1,nrow(sample_outlier_filter.df)))
sample_outlier_filter_nosgl.df.m=melt(sample_outlier_filter_nosgl.df,id.vars=c("sample_id","affected_status", "institution", "technology"))
detach(package:plyr)
library(dplyr)
#sample_outlier_filter_summary.df2= sample_outlier_filter.df2.m %>% group_by(variable, affected_status)%>% summarize(mean=mean(value, na.rm=T),sd=sd(value, na.rm=T))
sample_outlier_filter_summary_nosgl.df= sample_outlier_filter_nosgl.df.m %>% group_by(variable)%>% summarize(mean=mean(value, na.rm=T),sd=sd(value, na.rm=T))
sample_outlier_filter_summary_nosgl_affected_status.df= sample_outlier_filter_nosgl.df.m %>% group_by(variable, affected_status)%>% summarize(mean=mean(value, na.rm=T),sd=sd(value, na.rm=T))

#sample_outlier_filter_summary.df2$variable=factor(sample_outlier_filter_summary.df2$variable, levels=c("prop_init","prop_RV","prop_CADD"))
sample_outlier_filter_summary_nosgl.df$variable=factor(sample_outlier_filter_summary_nosgl.df$variable, levels=c("prop_pLI", "prop_RV","prop_RVpLI","prop_CADD","prop_combined"))
sample_outlier_filter_summary_nosgl_affected_status.df$variable=factor(sample_outlier_filter_summary_affected_status.df$variable, levels=c("prop_pLI", "prop_RV","prop_RVpLI","prop_CADD","prop_combined"))



sup_Fig_3d_nosgl=ggplot(sample_outlier_filter_summary_nosgl.df,aes(x=variable,y=mean))+ 
   geom_bar(position=position_dodge(),stat="identity", color="black", fill="#6D7993") + 
   geom_errorbar(aes(ymin=pmax(mean-sd,0), ymax=mean+sd), width=.2, position=position_dodge(.9))+
   #scale_fill_brewer(palette="Paired", name= "Affected status")+#RD_theme+
   labs(x="", y="Proportion of outlier genes")+RD_theme+ 
   scale_x_discrete(breaks=c("prop_init","prop_pLI","prop_RV","prop_RVpLI","prop_CADD","prop_combined"), labels=c("Zscore", "pLI", "RV","RV+pLI", "RV+CADD", "RV+pLI+CADD"))+theme(legend.position="")

ggsave('fig_3d_prop_candidates_withpLI.pdf', sup_Fig_3d_nosgl, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)



plot_affected_status_prop_candidates_nosgl=ggplot(sample_outlier_filter_summary_affected_status.df,aes(x=variable,y=mean, fill=affected_status))+ 
   geom_bar(position=position_dodge(),stat="identity", color="black") + 
   geom_errorbar(aes(ymin=pmax(mean-sd,0), ymax=mean+sd), width=.2, position=position_dodge(.9))+
   scale_fill_brewer(palette="Paired", name= "Affected status")+#RD_theme+
   labs(x="", y="Proportion of outlier genes")+RD_theme+ 
   scale_x_discrete(breaks=c("prop_init","prop_pLI","prop_RV","prop_RVpLI","prop_CADD","prop_combined"), labels=c("Zscore", "pLI", "RV","RV+pLI", "RV+CADD", "RV+pLI+CADD"))+theme(legend.position="bottom")

ggsave('sup_prop_candidates_withpLI_affected_status_nosgl.pdf', plot_affected_status_prop_candidates_nosgl, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)




# figure 3E link with disease?
# subset to neurology cases

# select Cases from neurology
neuro_samples=metadata %>% filter(disease_category=="Neurology")%>% filter(affected_status=="Case")%>% select(sample_id) %>% pull
hemato_samples=metadata %>% filter(disease_category=="Hematology")%>% filter(affected_status=="Case")%>% select(sample_id) %>% pull
opthalmo_samples=metadata %>% filter(disease_category=="Opthalmology")%>% filter(affected_status=="Case")%>% select(sample_id) %>% pull
cases=metadata %>% filter(affected_status=="Case")%>% select(sample_id) %>% pull
out_neuro=zscores_blood %>% filter(variable %in% neuro_samples) %>% filter(abs(value)>=2) %>% select(gene) %>% distinct


# Read in disease gene lists
disease_genes_dir='/srv/scratch/restricted/rare_diseases/data/candidate_gene_lists/'
gene_lists=list.files(path=disease_genes_dir, pattern='genelist.txt')
disease_lists=lapply(gene_lists, read_in_disease_lists, path_to_file=disease_genes_dir)
names(disease_lists)=unlist(strsplit(gene_lists, "_genelist.txt"))

# get ensembl names for all gene lists
disease_lists_ens=lapply(disease_lists,check_gene_names)


zscores_blood_neuro=zscores_blood %>% filter(variable %in% neuro_samples) %>% mutate(disease_gene=ifelse(gene %in% disease_lists_ens$Neurology,1,0))
zscores_blood_hemato=zscores_blood %>% filter(variable %in% hemato_samples) %>% mutate(disease_gene=ifelse(gene %in% disease_lists_ens$Hematology,1,0))
zscores_blood_opthal=zscores_blood %>% filter(variable %in% opthalmo_samples) %>% mutate(disease_gene=ifelse(gene %in% disease_lists_ens$Ophtalmology,1,0))
zscores_blood_omim=zscores_blood %>% filter(variable %in% cases) %>% mutate(disease_gene=ifelse(gene %in% as.character(disease_lists_ens$OMIM),1,0))
zscores_blood_omim2=zscores_blood %>% mutate(disease_gene=ifelse(gene %in% as.character(disease_lists_ens$OMIM),1,0))



test=zscores_blood_neuro %>% mutate(outlier=ifelse(abs(value)>=6, 1,0)) %>% select(gene, variable, disease_gene, outlier) %>% group_by(gene,variable) %>% summarize(disease_gene=max(disease_gene), outlier=max(outlier))%>% ungroup  %>% select(gene,disease_gene,outlier)
neuro.glmer=glmer(test$outlier ~test$disease_gene + (1|test$gene), family=binomial)


zscore_threshold=2:6



neuro.props= proportion.ratios(zscores_blood_neuro,zscore_threshold)
neuro.props$METHOD="splicing"
sig.splicing <- emp_ast(neuro.props$PVALUE)
neuro.props$SIG=sig.splicing


disease_colors=brewer.pal(12,"Paired")

neuro.props_plot.glm=ggplot(neuro.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color="gray") + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) +#ggtitle('Splicing Outlier')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    ylim(c(min(neuro.props$CI.LOW) - 0.1, max(neuro.props$CI.HIGH) + 0.1))+
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1, label=neuro.props$SIG), size=6) +
    annotate('text', x = neuro.props$THRESH-1, y =  neuro.props$CI.HIGH + 0.07,
             label = formatC(neuro.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme

ggsave('fig_3e_neurogenes_out_PIVUS_glm.pdf', neuro.props_plot.glm, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
ggsave('fig_3e_neurogenes_out_PIVUS_glm_poster.pdf', neuro.props_plot.glm, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=8, height=4)

neuro.props_plot.glm

# expression outlier


exp_zscores_file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/outliers_zscore_pair_02apr18.txt"
exp_zscores=as.data.frame(read.table(exp_zscores_file, header=T))


exp_zscores_blood_neuro=exp_zscores %>% filter(sample_id %in% neuro_samples) %>% mutate(disease_gene=ifelse(gene %in% disease_lists_ens$Neurology,1,0))


zscore_threshold=2:6

neuro.combined.props= proportion.ratios.combined(zscores_blood_neuro,exp_zscores_blood_neuro,zscore_threshold)
neuro.combined.props$METHOD="combined"

sig.combined <- emp_ast(neuro.combined.props$PVALUE)
neuro.combined.props$SIG=sig.combined

neuro.exp.props= proportion.ratios.exp(exp_zscores_blood_neuro,zscore_threshold)
neuro.exp.props$METHOD="expression"

sig.exp <- emp_ast(neuro.exp.props$PVALUE)
neuro.exp.props$SIG=sig.exp

#neuro.exp.signed.props=  proportion.ratios.exp.signed(exp_zscores_blood_neuro,zscore_threshold)


# combine all results
neuro.prop.all=rbind(neuro.props,neuro.exp.props,neuro.combined.props)

neuro.glm.plot=ggplot(neuro.prop.all, aes(x=as.factor(THRESH), y = LOG_ODDS, color=METHOD))+
	geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH))+ theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) +  
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    #ylim(c(min(neuro.combined.props$CI.LOW) - 0.1, max(neuro.combined.props$CI.HIGH) + 0.1))+
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1, label=neuro.prop.all$SIG), size=6) +
    annotate('text', x = neuro.prop.all$THRESH-1, y =  neuro.combined.props$CI.HIGH + 0.07,
             label = formatC(neuro.combined.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme

+geom_label_repel(aes(label = SIG,
    	color = factor(METHOD)), size = 3.5, show.legend = FALSE)
disease_colors=brewer.pal(12,"Paired")

neuro.props.combined_plot.glm=ggplot(neuro.combined.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Combined')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    ylim(c(min(neuro.combined.props$CI.LOW) - 0.1, max(neuro.combined.props$CI.HIGH) + 0.1))+
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1), label=sig.combined, size=6) +

    annotate('text', x = neuro.combined.props$THRESH-1, y =  neuro.combined.props$CI.HIGH + 0.07,
             label = formatC(neuro.combined.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme
neuro.props.combined_plot.glm




neuro.exp.props_plot.glm=ggplot(neuro.exp.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Expression Outlier')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1), label=sig.exp, size=6) +

    ylim(c(min(neuro.exp.props$CI.LOW) - 0.1, max(neuro.exp.props$CI.HIGH) + 0.1))+
    annotate('text', x = neuro.exp.props$THRESH-1, y =  neuro.exp.props$CI.HIGH + 0.07,
             label = formatC(neuro.exp.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme



neuro.exp.signed.props_plot.glm=ggplot(neuro.exp.signed.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Expression Outlier')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    ylim(c(min(neuro.exp.signed.props$CI.LOW) - 0.1, max(neuro.exp.signed.props$CI.HIGH) + 0.1))+
    annotate('text', x = neuro.exp.signed.props$THRESH-1, y =  neuro.exp.signed.props$CI.HIGH + 0.07,
             label = formatC(neuro.exp.signed.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme
ggsave('neuro.exp.signed.props_plot.glm.pdf', neuro.exp.signed.props_plot.glm, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

neuro.exp.signed.props_plot.glm
## combined_plot
disease_genes_enrich_plot=ggdraw()+draw_plot(neuro.props_plot, 0,0,1/3,1)+
	draw_plot(neuro.exp.props_plot, 1/3,0,1/3,1)+
	draw_plot(neuro.props.combined_plot,2/3,0,1/3,1)
disease_genes_enrich_plot2=ggdraw()+draw_plot(neuro.props_plot.glm, 0,0,1/3,1)+
	draw_plot(neuro.exp.props_plot.glm, 1/3,0,1/3,1)+
	draw_plot(neuro.props.combined_plot.glm,2/3,0,1/3,1)

ggsave('disease_genes_enrich_plot_glm.pdf', disease_genes_enrich_plot2, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=12, height=6)


## combined_plot
figure3_plot=ggdraw()+draw_plot(fig_3b_zscore_thres_poster, 0,1/3,1/2,0.4)+
	draw_plot(RV_out_plot_withsgl, 1/2,1/3,1/2,0.4)+
	draw_plot(filter_withsglt.case.plot, 0,0,1/2,1/3)+
	draw_plot(neuro.props_plot.glm, 1/2,0,1/2,1/3)+
	draw_plot_label(c('A', 'B', 'C','D', 'E'), c(0,0, 1/2,0,1/2), c(1,0.75,0.75,0.37,0.37), size = 20)

figure3_plot

pdf('/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/Figure3v7.pdf', w=14, h=14)
figure3_plot
dev.off()

# make a version without E
figure3_plot2=ggdraw()+draw_plot(fig_3b_zscore_thres_poster, 1/2,1/2,1/2,1/2)+
	draw_plot(RV_out_plot_withsgl, 0,0,1/2,1/2)+
	draw_plot(filter_withsglt.case.plot, 1/2,0,1/2,1/2)+
	draw_plot_label(c('A', 'B', 'C','D'), c(0,1/2,0, 1/2), c(1,1,1/2,1/2), size = 20)

pdf('/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/Figure3v8.pdf', w=12, h=12)
figure3_plot2
dev.off()

# make a vesrion with number of rare variants
# make a version without E
figure3_plot3=ggdraw()+draw_plot(fig_3b_zscore_thres_poster, 2/3,1/2,1/3,1/2)+
	draw_plot(RV_out_plot_withsgl, 0,0,1/3,1/2)+
	draw_plot(filter_withsglt.case.plot, 1/3,0,1/3,1/2)+
	draw_plot(RV_filters_plot, 2/3,0,1/3,1/2)+
	draw_plot_label(c('A', 'B', 'C','D', 'E'), c(0,2/3,0, 1/3,2/3), c(1,1,1/2,1/2, 1/2), size = 20)

pdf('/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/Figure3v8.pdf', w=12, h=12)
figure3_plot3
dev.off()



# Investigation of known causal genes properties
zscores_blood_max=zscores_blood %>% group_by(gene, variable) %>% summarize(max_splZ=max(absZ))
causal_genes=metadata%>% 
	filter(resolved_case=="yes") %>% 
	filter(affected_status=="Case") %>% 
	select(sample_id,causal_gene, variant_data, mutation_chr,mutation_bp, HPO_terms_ID)%>%  
	mutate(causal=str_extract(causal_gene, "(^[A-Z]+)[[0-9]+]?[[A-Z]]?")) %>%
	left_join(grch37, by=c("causal"="symbol")) %>% 
	select(sample_id,causal,ensgene,description,variant_data, mutation_chr,mutation_bp,HPO_terms_ID) %>% 
	distinct %>% 
	left_join(exac,by="ensgene") %>% 
	select(sample_id,causal, ensgene,syn_z,mis_z, lof_z, pLI,variant_data, mutation_chr,mutation_bp,HPO_terms_ID)%>% 
	left_join(exp_zscores, by=c("ensgene"="gene", "sample_id"="sample_id"))%>% 
	rename("exp_zscore"="zscore")%>% 
	distinct %>%
	left_join(zscores_blood_max, by=c("ensgene"="gene", "sample_id"="variable"))%>%
	left_join(RV_outlier_filt_withsgl, by=c("sample_id"="variable", "ensgene"="gene_name", "mutation_chr"="chr_var", "mutation_bp"="end_var"))%>%
	select(sample_id,HPO_terms_ID,causal,ensgene,syn_z,mis_z, lof_z, pLI,variant_data, mutation_chr,mutation_bp,exp_zscore,max_splZ,gnomAD_AF_filt_withsgl,gnomAD_AF_filt_nosgl,phred_cadd)%>% distinct


# generates results without RV
candidates_no_gen_ext=lapply(cases,get_hpo_comp_nogen,zscores_blood,metadata, gene_to_pheno,pheno_annot,HPO_extended)
names(candidates_no_gen_ext)=cases

candidates_zscore_hpo_ext=do.call("rbind", candidates_no_gen_ext)
candidates_zscore_hpo_ext=candidates_zscore_hpo_ext %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, institution_id, institution)
write.table(candidates_zscore_hpo_ext, "splicingoutliers_hpo_candidates.tsv", sep="\t", row.names=F, append=F, quote=F)

# generate results with RV
#potential_can=lapply(cases_withGen,get_hpo_comp,RV_outlier_filt_withsgl,metadata, gene_to_pheno,pheno_annot, 10,HPO_extended)
potential_can=lapply(cases_withGen,get_hpo_comp_window,RV_outlier_filt_withsgl,metadata, gene_to_pheno,pheno_annot, 10,HPO_extended)
names(potential_can)=cases_withGen

candidates_zscore_hpo_RV=do.call("rbind", potential_can)
candidates_zscore_hpo_RV=candidates_zscore_hpo_RV %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, indv_id, institution)
write.table(candidates_zscore_hpo_RV, "splicingoutliers_hpo_candidates_RV.tsv", sep="\t", row.names=F, append=F, quote=F)

# generate results with RV no CADD score threshold
#potential_can=lapply(cases_withGen,get_hpo_comp,RV_outlier_filt_withsgl,metadata, gene_to_pheno,pheno_annot, 10,HPO_extended)
potential_can_nocadd=lapply(cases_withGen,get_hpo_comp_window,RV_outlier_filt_withsgl,metadata, gene_to_pheno,pheno_annot, 0,HPO_extended)
names(potential_can_nocadd)=cases_withGen

candidates_zscore_hpo_RV_nocadd=do.call("rbind", potential_can_nocadd)
candidates_zscore_hpo_RV_nocadd=candidates_zscore_hpo_RV_nocadd %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, indv_id, institution)
write.table(candidates_zscore_hpo_RV_nocadd, "splicingoutliers_hpo_candidates_RV_nocadd.tsv", sep="\t", row.names=F, append=F, quote=F)



candidates=zscores_blood %>% filter(variable=="RD047") %>% filter(absZ>=2) %>% 
	left_join(grch37, by=c("gene"="ensgene")) %>%
		select(gene, symbol, entrez, absZ)
hpo_comb=candidates %>% left_join(gene_to_pheno, by=c("entrez"="entrez-gene-id")) %>% distinct %>% left_join(pheno_annot,by=c("DiseaseId"="database_INDEX")) %>% distinct %>% mutate(hpo_in_input=ifelse(HPO_terms_ID %in% HPOs[["RD064"]], 1, 0)) %>% select(symbol, hpo_in_input) %>% distinct %>%group_by(symbol)%>% summarise(overlap_hpo=sum(hpo_in_input)) %>% filter(overlap_hpo>=1)



potential_can=lapply(cases_withGen,get_hpo_comp,RV_outlier_filt_withsgl,metadata, gene_to_pheno,pheno_annot, 10,HPO_extended)
names(potential_can)=cases_withGen

candidates_zscore_hpo_RV=do.call("rbind", potential_can)
candidates_zscore_hpo_RV=candidates_zscore_hpo_RV %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, indv_id, institution)
write.table(candidates_zscore_hpo_RV, "splicingoutliers_hpo_candidates_RV.tsv", sep="\t", row.names=F, append=F, quote=F)

candidates_no_gen=lapply(cases,get_hpo_comp_nogen,zscores_blood,metadata, gene_to_pheno,pheno_annot,HPOs)
names(candidates_no_gen)=cases

candidates_zscore_hpo=do.call("rbind", candidates_no_gen)
candidates_zscore_hpo=candidates_zscore_hpo %>% arrange(sample)%>% left_join(metadata, by=c("sample"="sample_id")) %>% select( symbol,overlap_hpo,    max_z ,sample, indv_id, institution)

write.table(candidates_zscore_hpo, "splicingoutliers_hpo_candidates.tsv", sep="\t", row.names=F, append=F, quote=F)



save(zscores_blood,RV_outlier_filt_withsgl,HPO_extended,sample_with_data,metadata,affected_status_df,candidates_zscore_hpo_ext ,candidates_zscore_hpo_RV,file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/splicing_outlier_files.Rdata")


# max Z
max_z=zscores_blood %>% filter(status=="Case") %>%select(variable,gene, chr, junc_start, junc_end,absZ)%>% group_by(variable) %>%  filter(absZ==max(absZ, na.rm=T)) %>% as.data.frame %>% ungroup %>% distinct(variable,gene, chr, junc_start, junc_end,absZ) %>% left_join(grch37, by=c("gene"="ensgene"))
write.table(max_z, "splicingoutliers_maxZ_candidates.tsv", sep="\t", row.names=F, append=F, quote=F)


top10_z=zscores_blood %>% filter(status=="Case") %>%select(variable,gene, chr, junc_start, junc_end,absZ)%>% group_by(variable) %>%  top_n(10,absZ) %>% as.data.frame %>% ungroup %>% distinct(variable,gene, chr, junc_start, junc_end,absZ) %>% left_join(grch37, by=c("gene"="ensgene"))
write.table(top10_z, "splicingoutliers_top10Z_candidates.tsv", sep="\t", row.names=F, append=F, quote=F)

