#!/bin/R
#April 2017

# LF

# This scripts performs analyses for influence of differet filters on the number of expression outlier gene candidates for Figure 2
#Libraries
library(ggplot2)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(sva)
library(cowplot)
library(purrr)
library(broom)
library(reshape2)
library(ggfortify)
library(gridExtra)
library(data.table)
library(annotables)
library(corrplot)
library(RColorBrewer)
library(qvalue) 
library(ggpubr)


## FUNCTIONS


rm(list=ls())

source('/users/lfresard/repos/rare_disease/scripts/manuscript_analyses/Figures_source.R')


fsize=20
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize),
	axis.text.y= element_text(size=fsize), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.1),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=9),
	panel.border=element_blank()) 



## MAIN
# Get metadata
metadata = "/srv/scratch/restricted/rare_diseases/data/metadata/2018_06_12_Rare_Disease_Metadata.tsv"
metadata = read_tsv(metadata) %>% filter(in_freeze=="yes") %>% filter(source_of_RNA=="Blood")

# get list of all RD samples
RD_samples=metadata$sample_id

# get list of samples with genetic data
sample_with_data=metadata %>% filter(variant_data=="exome" | variant_data=="genome")%>% select(sample_id) %>% pull

# read in exac data
exac = as.data.frame(read.table('/users/lfresard/NumberOfIndividuals_Impact/enrichment_disease-conservation_database/annotations/EXAC/forweb_cleaned_exac_r03_march16_z_data_pLI.txt', sep = '\t', stringsAsFactors = F, header = T))
exac=exac %>% inner_join(grch37, by=c("gene"="symbol")) %>% select(ensgene,syn_z, mis_z, lof_z, pLI) 

# read in affected status data
#affected_status_df=data.frame(sample=metadata$sample_id,status=metadata$affected_status)
affected_status_df=read.table('/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/splicing/for_freeze/ifrun9/sample_affected_status_freeze_RD_PIVUS.tsv', header=F)
colnames(affected_status_df)=c('sample', 'status')


# read in phenotypic data
# Read in annotation files linking HPO terms genes and phenotypes
gene_to_pheno="/srv/scratch/restricted/rare_diseases/data/metadata/gene_to_phenotype.tsv"
gene_to_pheno = read_tsv(gene_to_pheno)
colnames(gene_to_pheno)=c("entrez", "symbol", "DiseaseId")

pheno_annot="/srv/scratch/restricted/rare_diseases/data/metadata/phenotype_annotation.tsv"
pheno_annot=read_tsv(pheno_annot)
pheno_annot=pheno_annot %>% select(HPO_terms_ID,database_INDEX, number_unknown)

# Read in and process alternative HPO files (generated by parse_hpoterms.py)
parent_terms="/srv/scratch/restricted/rare_diseases/data/metadata/parent_terms_hpo_2018_03_16.txt"
child_terms="/srv/scratch/restricted/rare_diseases/data/metadata/child_terms_hpo_2018_03_16.txt"
altid_terms="/srv/scratch/restricted/rare_diseases/data/metadata/altid_terms_hpo_2018_03_16.txt"

altid_terms=as.data.frame(read.table(altid_terms, header=F, sep="\t"))
colnames(altid_terms)=c("HPO_ID", "ALT_ID")
child_terms=as.data.frame(read.table(child_terms, header=F, sep="\t"))
colnames(child_terms)=c("HPO_ID", "ALT_ID")
parent_terms=as.data.frame(read.table(parent_terms, header=F, sep="\t"))
colnames(parent_terms)=c("HPO_ID", "ALT_ID")


# generates list of HPO alternative IDs from the input files
alt_hpo_ids_list=lapply(altid_terms$HPO_ID,get_hpo_alt_id, altid_terms=altid_terms)
names(alt_hpo_ids_list)=altid_terms$HPO_ID
child_hpo_ids_list=lapply(child_terms$HPO_ID,get_hpo_alt_id, altid_terms=child_terms)
names(child_hpo_ids_list)=child_terms$HPO_ID
parent_hpo_ids_list=lapply(parent_terms$HPO_ID,get_hpo_alt_id, altid_terms=parent_terms)
names(parent_hpo_ids_list)=parent_terms$HPO_ID

# For every ID associated with the disease, get list of alternative Ids, child ids and parent ids
# make a list for each case with HPO terms

cases=metadata %>% filter(affected_status=="Case") %>% select(sample_id) %>% pull

# Get list of HPO terms from metadata file
HPOs=lapply(cases,get_list_hpo, metadata)
names(HPOs)=cases
cases_withGen=metadata %>% filter(variant_data !="none", affected_status=="Case")%>% select(sample_id) %>% pull

# Generate extended list of HPO terms for each sample
HPO_extended=lapply(cases, process_sample_ids_hpo, HPO_list=HPOs)
names(HPO_extended)=cases


# expression outlier data
exp_outlier="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/outliers_zscore_pair_spline_05apr18.txt"
exp_outlier=as.data.frame(read.table(exp_outlier, header=T))


# expression outlier with rare variant (window method)
RV_outlier_exp_10kb="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/outliers_rare_var_combined_10kb.txt"
RV_outlier_exp_10kb=as.data.frame(read.table(RV_outlier_exp_10kb, sep="\t", header=F))
colnames(RV_outlier_exp_10kb)=c("sample_id", "gene", "expressionZScore", "variantPos", "gnomadMAF", "raw_cadd", "phred_cadd", "variant_gene_pos")

RV_outlier_exp_10kb=RV_outlier_exp_10kb %>% left_join(grch37, by=c("gene"="ensgene"), all.x=T)%>%select(sample_id,gene,expressionZScore,gnomadMAF,phred_cadd,variant_gene_pos,variantPos, start,end)

# get distance of variant from gene
RV_outlier_exp_10kb=RV_outlier_exp_10kb%>% mutate(distance=ifelse((variantPos<=start&variantPos<=end)|(variantPos>=start&variantPos>=end),start-variantPos,0))

# make a column with singleton included or expluded form the analysis (window method)
RV_outlier_exp_10kb =RV_outlier_exp_10kb %>% mutate(gnomAD_AF_withsgl = replace(gnomadMAF, gnomadMAF==".", 0))
RV_outlier_exp_10kb =RV_outlier_exp_10kb %>% mutate(gnomAD_AF_nosgl = replace(gnomadMAF, gnomadMAF==".", NA))


# format variant data so that only the max observed allele frequency is kept for variant where several frequencies are available (window method)
RV_outlier_exp_10kb$gnomAD_AF_filt_withsgl=sapply(RV_outlier_exp_10kb$gnomAD_AF_withsgl,process_variant)
RV_outlier_exp_10kb$gnomAD_AF_filt_nosgl=sapply(RV_outlier_exp_10kb$gnomAD_AF_nosgl,process_variant)


#filter for RV with MAF<=0.1% (window method)
maf=0.001
	# Excluding singletons
RV_outlier_exp_filt_nosgl_10kb=RV_outlier_exp_10kb %>% filter(!is.na(gnomAD_AF_filt_nosgl)) %>% filter(gnomAD_AF_filt_nosgl <=maf)
	# Including singletons
RV_outlier_exp_filt_withsgl_10kb=RV_outlier_exp_10kb %>% filter(!is.na(gnomAD_AF_filt_withsgl)) %>% filter(gnomAD_AF_filt_withsgl <=maf)


# Read in RIVER scores
river_file="/mnt/lab_data/montgomery/xli6/udn/river_score/UDN_RIVERscore_ByGene.txt"
river=as.data.frame(read.table(river_file, sep="\t", header=T))
	#get list of genes with RIVER score >=0.85 for each sample
gene_river_list=sapply(sample_with_data,get_genes_with_highriver,river)



# Read in ASE data
ase_file="/srv/scratch/restricted/rare_diseases/data/ase/June13/final_merged/ensembleids.ase.filtered.txt"
ase=read_tsv(ase_file, col_names=TRUE)
colnames(ase)=c("sample_id", colnames(ase)[2:length(colnames(ase))])

	# transform ase data 
ase=ase %>% mutate(ensgene=str_extract(Gene, "ENSG[0-9]+")) %>% select(sample_id,Chr,POS, "RefRatio" , "TotalReads","ensgene")%>% filter (! duplicated(ensgene))%>% mutate(RefRatio_trans=ifelse(RefRatio<=0.5,1-RefRatio,RefRatio))%>% filter (! RefRatio_trans==1)
ase=ase%>% group_by(ensgene)%>% mutate(max_RefRatio_trans=max(RefRatio_trans,na.rm=T), median_RefRatio_trans=median(RefRatio_trans, na.rm=T))%>% ungroup

#hist_ase=hist(ase$median_RefRatio_trans)
#ase1=ase %>% filter(median_RefRatio_trans==1)
#hist_ase1=hist(ase1$TotalReads, breaks=30)

	# get list of genes with ASE for each sample
gene_ase_list=sapply(sample_with_data,get_genes_with_ase,ase)

# filter for outlier genes with ASE

## Influence of filtering on expression outlier results
zscore_threshold=2
distance_threshold=10000
cadd=10
pli=0.9





sample_exp_outlier_filter_withsgl_10kb_up.df=data.frame(sample_id=sample_with_data,
	"raw"=sapply(sample_with_data, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=0, direction="under"),
	"raw.Lof"=sapply(sample_with_data, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=pli, direction="under"),
	"raw.HPO"=sapply(sample_with_data,get_hpo_match_nb_exp,outlier_file=exp_outlier,HPOs_list=HPO_extended, pLI_thre=0, direction="under"),
	"raw.ASE"=sapply(sample_with_data, get_outlier_ase_genes, threshold=zscore_threshold, outlier_df=exp_outlier, gene_ase_list=gene_ase_list,pLI_thre=0, direction="under"),
	"raw.river"=sapply(sample_with_data, get_outlier_river_genes, threshold=zscore_threshold, outlier_df=exp_outlier, gene_river_list=gene_river_list,pLI_thre=0, direction="under"),
	"RV"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=0, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.LoF.intolerant"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=0, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),
	"RV.CADD"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.ASE"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb_ase,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_ase_list, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.RIVER"=sapply(sample_with_data, get_outlier_genes_RV_exp_10kb_ase,  outlier_df.m=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_river_list, CADD=cadd, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD.pLI"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),
	"RV.HPO.CADD.ASE"=sapply(sample_with_data, get_hpo_match_nb_RV_exp_10kb_ase,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb,gene_ase_list=gene_ase_list,HPOs_list=HPO_extended, cadd_thres=cadd, pLI_thre=pli, direction="under", myvariant_gene_pos="upstream"),

	institution=sapply(sample_with_data,get_institution, metadata=metadata),
	affected_status=sapply(sample_with_data,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(sample_with_data,get_technology, metadata=metadata))


save("sample_exp_outlier_filter_withsgl_10kb_up.df", file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/expression_filters_genes.RData")








# function that generate the number of outlier genes with RV within 20bp of juction for which at list one HPO terms match with symptomes, CADD and PLI scores adjustable
get_hpo_match_RV_exp_10kb=function(sample,outlier_RV, HPOs_list, cadd_thres, pLI_thre, direction,myvariant_gene_pos){
	if (direction=="over"){
		if(cadd_thres==0){
	gene_match=outlier_RV %>% filter(sample_id==sample) %>% filter(expressionZScore>=2) %>% 
		filter(variant_gene_pos==myvariant_gene_pos)%>%
		filter(as.numeric(as.character(phred_cadd))>=cadd_thres|phred_cadd==".")%>%
		left_join(exac,by=c("gene"="ensgene"), all.x=T) %>% distinct %>% 
		filter(pLI>=pLI_thre) %>%
		left_join(grch37, by=c("gene"="ensgene")) %>%
		select(gene, symbol, entrez, expressionZScore) %>% 
		left_join(gene_to_pheno, by=c("entrez","symbol"))%>% distinct %>% 
		left_join(pheno_annot,by=c("DiseaseId"="database_INDEX")) %>% distinct %>%
		mutate(hpo_in_input=ifelse(HPO_terms_ID %in% HPOs_list[[sample]], 1, 0)) %>%
		select(gene, symbol, expressionZScore, hpo_in_input) %>% distinct %>%group_by(gene,symbol, expressionZScore)%>% summarise(overlap_hpo=sum(hpo_in_input)) %>% 
 		filter(overlap_hpo>=1) %>% ungroup %>% unique%>% filter (! duplicated(gene))%>% select(gene) %>% pull 
	}else{
	gene_match=outlier_RV %>% filter(sample_id==sample) %>% filter(expressionZScore>=2) %>% 
		filter(variant_gene_pos==myvariant_gene_pos)%>%
		filter(as.numeric(as.character(phred_cadd))>=cadd_thres)%>%
		left_join(exac,by=c("gene"="ensgene"), all.x=T) %>% distinct %>% 
		filter(pLI>=pLI_thre) %>%
		left_join(grch37, by=c("gene"="ensgene")) %>%
		select(gene, symbol, entrez, expressionZScore) %>% 
		left_join(gene_to_pheno, by=c("entrez","symbol"))%>% distinct %>% 
		left_join(pheno_annot,by=c("DiseaseId"="database_INDEX")) %>% distinct %>%
		mutate(hpo_in_input=ifelse(HPO_terms_ID %in% HPOs_list[[sample]], 1, 0)) %>%
		select(gene, symbol, expressionZScore, hpo_in_input) %>% distinct %>%group_by(gene,symbol, expressionZScore)%>% summarise(overlap_hpo=sum(hpo_in_input)) %>% 
 		filter(overlap_hpo>=1) %>% ungroup %>% unique%>% filter (! duplicated(gene)) %>%  select(gene) %>%pull 


	}
	}
	if (direction=="under"){
			if(cadd_thres==0){
	gene_match=outlier_RV %>% filter(sample_id==sample) %>% filter(expressionZScore<=-2) %>% 
		filter(variant_gene_pos==myvariant_gene_pos)%>%
		filter(as.numeric(as.character(phred_cadd))>=cadd_thres|phred_cadd==".")%>%
		left_join(exac,by=c("gene"="ensgene"), all.x=T) %>% distinct %>% 
		filter(pLI>=pLI_thre) %>%
		left_join(grch37, by=c("gene"="ensgene")) %>%
		select(gene, symbol, entrez, expressionZScore) %>% 
		left_join(gene_to_pheno, by=c("entrez","symbol"))%>% distinct %>% 
		left_join(pheno_annot,by=c("DiseaseId"="database_INDEX")) %>% distinct %>%
		mutate(hpo_in_input=ifelse(HPO_terms_ID %in% HPOs_list[[sample]], 1, 0)) %>%
		select(gene, symbol, expressionZScore, hpo_in_input) %>% distinct %>%group_by(gene,symbol, expressionZScore)%>% summarise(overlap_hpo=sum(hpo_in_input)) %>% 
 		filter(overlap_hpo>=1) %>% ungroup %>% unique%>% filter (! duplicated(gene)) %>%  select(gene) %>%pull 
	}else{
	gene_match=outlier_RV %>% filter(sample_id==sample) %>% filter(expressionZScore<=-2) %>% 
		filter(variant_gene_pos==myvariant_gene_pos)%>%
		filter(as.numeric(as.character(phred_cadd))>=cadd_thres)%>%
		left_join(exac,by=c("gene"="ensgene"), all.x=T) %>% distinct %>% 
		filter(pLI>=pLI_thre) %>%
		left_join(grch37, by=c("gene"="ensgene")) %>%
		select(gene, symbol, entrez, expressionZScore) %>% 
		left_join(gene_to_pheno, by=c("entrez","symbol"))%>% distinct %>% 
		left_join(pheno_annot,by=c("DiseaseId"="database_INDEX")) %>% distinct %>%
		mutate(hpo_in_input=ifelse(HPO_terms_ID %in% HPOs_list[[sample]], 1, 0)) %>%
		select(gene, symbol, expressionZScore, hpo_in_input) %>% distinct %>%group_by(gene,symbol, expressionZScore)%>% summarise(overlap_hpo=sum(hpo_in_input)) %>% 
 		filter(overlap_hpo>=1) %>% ungroup %>% unique%>% filter (! duplicated(gene)) %>%  select(gene) %>%pull 

	}}
 	return(gene_match)
}


# 1/ underexpression upstream cadd10 
RV.HPO.CADD_under_upstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="under", myvariant_gene_pos="upstream")
names(RV.HPO.CADD_under_upstream)=sample_with_data
RV.HPO.CADD_under_upstream=compact(RV.HPO.CADD_under_upstream)

RV.HPO.CADD_under_upstream=Map(cbind, RV.HPO.CADD_under_upstream, sample =names(RV.HPO.CADD_under_upstream))
RV.HPO.CADD_under_upstream_df=as.data.frame(do.call(rbind, RV.HPO.CADD_under_upstream))
RV.HPO.CADD_under_upstream_df$outlier="underexpression"
RV.HPO.CADD_under_upstream_df$cadd_thres=10
RV.HPO.CADD_under_upstream_df$variant_pos="upstream_and_gene"

# 2/ overexpression upstream cadd10 
RV.HPO.CADD_over_upstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="over", myvariant_gene_pos="upstream")
names(RV.HPO.CADD_over_upstream)=sample_with_data
RV.HPO.CADD_over_upstream=compact(RV.HPO.CADD_over_upstream)

RV.HPO.CADD_over_upstream=Map(cbind, RV.HPO.CADD_over_upstream, sample =names(RV.HPO.CADD_over_upstream))
RV.HPO.CADD_over_upstream_df=as.data.frame(do.call(rbind, RV.HPO.CADD_over_upstream))
RV.HPO.CADD_over_upstream_df$outlier="overrexpression"
RV.HPO.CADD_over_upstream_df$cadd_thres=10
RV.HPO.CADD_over_upstream_df$variant_pos="upstream_and_gene"

# 3/ underexpression downstream cadd10 
RV.HPO.CADD_under_downstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="under", myvariant_gene_pos="downstream")
names(RV.HPO.CADD_under_downstream)=sample_with_data
RV.HPO.CADD_under_downstream=compact(RV.HPO.CADD_under_downstream)

RV.HPO.CADD_under_downstream=Map(cbind, RV.HPO.CADD_under_downstream, sample =names(RV.HPO.CADD_under_downstream))
RV.HPO.CADD_under_downstream_df=as.data.frame(do.call(rbind, RV.HPO.CADD_under_downstream))
RV.HPO.CADD_under_downstream_df$outlier="underexpression"
RV.HPO.CADD_under_downstream_df$cadd_thres=10
RV.HPO.CADD_under_downstream_df$variant_pos="downstream_and_gene"

# 4/ overexpression downstream cadd10 
RV.HPO.CADD_over_downstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=10, pLI_thre=0, direction="over", myvariant_gene_pos="downstream")
names(RV.HPO.CADD_over_downstream)=sample_with_data
RV.HPO.CADD_over_downstream=compact(RV.HPO.CADD_over_downstream)

RV.HPO.CADD_over_downstream=Map(cbind, RV.HPO.CADD_over_downstream, sample =names(RV.HPO.CADD_over_downstream))
RV.HPO.CADD_over_downstream_df=as.data.frame(do.call(rbind, RV.HPO.CADD_over_downstream))
RV.HPO.CADD_over_downstream_df$outlier="overexpression"
RV.HPO.CADD_over_downstream_df$cadd_thres=10
RV.HPO.CADD_over_downstream_df$variant_pos="downstream_and_gene"



# 5/ underexpression upstream cadd0 
RV.HPO_under_upstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="under", myvariant_gene_pos="upstream")
names(RV.HPO_under_upstream)=sample_with_data
RV.HPO_under_upstream=compact(RV.HPO_under_upstream)

RV.HPO_under_upstream=Map(cbind, RV.HPO_under_upstream, sample =names(RV.HPO_under_upstream))
RV.HPO_under_upstream_df=as.data.frame(do.call(rbind, RV.HPO_under_upstream))
RV.HPO_under_upstream_df$outlier="underexpression"
RV.HPO_under_upstream_df$cadd_thres=0
RV.HPO_under_upstream_df$variant_pos="upstream_and_gene"

# 6/ overexpression upstream cadd0 
RV.HPO_over_upstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="over", myvariant_gene_pos="upstream")
names(RV.HPO_over_upstream)=sample_with_data
RV.HPO_over_upstream=compact(RV.HPO_over_upstream)

RV.HPO_over_upstream=Map(cbind, RV.HPO_over_upstream, sample =names(RV.HPO_over_upstream))
RV.HPO_over_upstream_df=as.data.frame(do.call(rbind, RV.HPO_over_upstream))
RV.HPO_over_upstream_df$outlier="overrexpression"
RV.HPO_over_upstream_df$cadd_thres=0
RV.HPO_over_upstream_df$variant_pos="upstream_and_gene"

# 7/ underexpression downstream cadd0 
RV.HPO_under_downstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="under", myvariant_gene_pos="downstream")
names(RV.HPO_under_downstream)=sample_with_data
RV.HPO_under_downstream=compact(RV.HPO_under_downstream)

RV.HPO_under_downstream=Map(cbind, RV.HPO_under_downstream, sample =names(RV.HPO_under_downstream))
RV.HPO_under_downstream_df=as.data.frame(do.call(rbind, RV.HPO_under_downstream))
RV.HPO_under_downstream_df$outlier="underexpression"
RV.HPO_under_downstream_df$cadd_thres=0
RV.HPO_under_downstream_df$variant_pos="downstream_and_gene"

# 8/ overexpression downstream cadd0 
RV.HPO_over_downstream=lapply(sample_with_data, get_hpo_match_RV_exp_10kb,  outlier_RV=RV_outlier_exp_filt_withsgl_10kb, HPOs_list=HPO_extended,cadd_thres=0, pLI_thre=0, direction="over", myvariant_gene_pos="downstream")
names(RV.HPO_over_downstream)=sample_with_data
RV.HPO_over_downstream=compact(RV.HPO_over_downstream)

RV.HPO_over_downstream=Map(cbind, RV.HPO_over_downstream, sample =names(RV.HPO_over_downstream))
RV.HPO_over_downstream_df=as.data.frame(do.call(rbind, RV.HPO_over_downstream))
RV.HPO_over_downstream_df$outlier="overexpression"
RV.HPO_over_downstream_df$cadd_thres=0
RV.HPO_over_downstream_df$variant_pos="downstream_and_gene"



candidates_summary=rbind(RV.HPO_over_downstream_df,RV.HPO_under_downstream_df,RV.HPO_over_upstream_df,RV.HPO_under_upstream_df,RV.HPO.CADD_over_downstream_df,RV.HPO.CADD_under_downstream_df,RV.HPO.CADD_over_upstream_df,RV.HPO.CADD_under_upstream_df)
colnames(candidates_summary)=c("ensgene","sample", "outlier_type","cadd_threshold","variant_pos")

write.table(candidates_summary,"/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/expression_outlierscandidate_genes.txt", , col.names=T, row.names=F, quote=F, sep="\t", append=F )


# make a plot with proportions of candidates 

sample_exp_outlier_filter_withsgl_10kb_up.df= sample_exp_outlier_filter_withsgl_10kb_up.df %>% filter(raw!=0)
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.Lof/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_HPO=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.HPO/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RIVER=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.river/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$raw.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV=sample_exp_outlier_filter_withsgl_10kb_up.df$RV/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.LoF.intolerant/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_CADD=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.CADD/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.RIVER=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.RIVER/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD.pLI=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD.pLI/sample_exp_outlier_filter_withsgl_10kb_up.df$raw
sample_exp_outlier_filter_withsgl_10kb_up.df$prop_RV.HPO.CADD.ASE=sample_exp_outlier_filter_withsgl_10kb_up.df$RV.HPO.CADD.ASE/sample_exp_outlier_filter_withsgl_10kb_up.df$raw

sample_exp_outlier_filter_withsgl_10kb_up.df=sample_exp_outlier_filter_withsgl_10kb_up.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_RIVER, prop_ASE, prop_HPO,prop_RV,prop_RV.pLI,prop_CADD,prop_RV.RIVER, prop_RV.HPO, prop_RV.ASE,prop_RV.HPO.CADD,prop_RV.HPO.CADD.pLI,prop_RV.HPO.CADD.ASE)

#melt DF
sample_exp_outlier_filter_withsgl_10kb_up.df.m=melt(sample_exp_outlier_filter_withsgl_10kb_up.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_exp_outlier_filter_withsgl_10kb_up.df.m.cases=sample_exp_outlier_filter_withsgl_10kb_up.df.m %>% filter(affected_status=="Case")

# plot results
filter_withsglt.exp.case.10kb.plot=ggplot(sample_exp_outlier_filter_withsgl_10kb_up.df.m.cases, aes(x=variable, y=value, fill=variable) )+ 
	geom_boxplot(color="black", notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 13), breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c(expression("pLI " >="0.9"), expression("RIVER " >= "0.85"), "ASE", "HPO match", expression("Rare variant within 10kb","1 + 5","5 + CADD score  " >= "10"), "2 + 5", "4 + 5", "3 + 5", "4 + 7", "1 + 2 + 5 + 7", "2 + 4 + 5 + 7"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c("1", "2", "3","4","5", "6", "7", "8", "9", "10", "11", "12", "13"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56,57,58,59,60,61))))+theme(legend.position = c(0.8,0.7))
filter_withsglt.exp.case.10kb.plot

ggsave('fig_2_prop_candidates_filters_10kb_upstream.pdf', filter_withsglt.exp.case.10kb.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)
ggsave('fig_2_prop_candidates_filters_10kb_upstream.pdf', filter_withsglt.exp.case.10kb.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=8, height=8)
save(filter_withsglt.exp.case.10kb.plot,file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/exp_filters.Rdata")
filter_withsglt.exp.case.10kb_techno.plot=ggplot(sample_exp_outlier_filter_withsgl_10kb_up.df.m.cases, aes(x=variable, y=value, fill=variable, color=technology) )+ 
	geom_boxplot( notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 13), breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c(expression("pLI " >="0.9"), expression("RIVER " >= "0.85"), "ASE", "HPO match", expression("Rare variant within 10kb","1 + 5","5 + CADD score  " >= "10"), "2 + 5", "4 + 5", "3 + 5", "4 + 7", "1 + 11", "3 + 12"), name="Filter")+
	scale_color_manual(values=c("#003366", "#CCCC99"))+ 
	scale_x_discrete(breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c("1", "2", "3","4","5", "6", "7", "8", "9", "10", "11", "12", "13"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56,57,58,59,60,61))), colour=guide_legend(override.aes = list(size = 4,shape =c(1,1)), fill=c("#003366", "#CCCC99")))+theme(legend.position = c(0.8,0.5))
filter_withsglt.exp.case.10kb_techno.plot

ggsave('sup_prop_candidates_filters_10kb_upstream_techno.pdf', filter_withsglt.exp.case.10kb_techno.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

filter_withsglt.exp.case.10kb_affected_status.plot=ggplot(sample_exp_outlier_filter_withsgl_10kb_up.df.m, aes(x=variable, y=value, fill=variable, color=affected_status) )+ 
	geom_boxplot( notch=F, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 13), breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c(expression("pLI " >="0.9"), expression("RIVER " >= "0.85"), "ASE", "HPO match", expression("Rare variant within 10kb","1 + 5","5 + CADD score  " >= "10"), "2 + 5", "4 + 5", "3 + 5", "4 + 7", "1 + 11", "3 + 12"), name="Filter")+
	scale_color_brewer(palette="Paired",name="Affected status")+ 
	scale_x_discrete(breaks=c("prop_pLI","prop_RIVER","prop_ASE", "prop_HPO","prop_RV", "prop_RV.pLI", "prop_CADD","prop_RV.RIVER","prop_RV.HPO", "prop_RV.ASE", "prop_RV.HPO.CADD","prop_RV.HPO.CADD.pLI","prop_RV.HPO.CADD.ASE"),
		labels=c("1", "2", "3","4","5", "6", "7", "8", "9", "10", "11", "12", "13"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51,52,53,54,55,56,57,58,59,60,61))), colour=guide_legend(override.aes = list(size = 4,shape =c(1,1)), fill=c("#003366", "#CCCC99")))+theme(legend.position = c(0.8,0.5))
filter_withsglt.exp.case.10kb_affected_status.plot

ggsave('sup_prop_candidates_filters_10kb_upstream_affected_status.pdf', filter_withsglt.exp.case.10kb_affected_status.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)

# make summary of means per filter
exp_filters_summary=sample_exp_outlier_filter_withsgl_10kb_up.df.m%>% group_by(variable) %>% summarize(mean=mean(value))

# write results
write.table(exp_filters_summary,"/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/exp_filters_summary.txt", , col.names=T, row.names=F, quote=F, sep="\t", append=F )

sample_outlier_filter_nogen_exp_under.df=data.frame(sample_id=RD_samples,
	"raw"=sapply(RD_samples, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=0, direction="under"),
	"raw.Lof"=sapply(RD_samples, get_outlier_genes_exp, threshold=zscore_threshold, outlier_df.m=exp_outlier, pLI_thre=pli, direction="under"),
	"raw.HPO"=sapply(RD_samples,get_hpo_match_nb_exp,outlier_file=exp_outlier,HPOs_list=HPO_extended, pLI_thre=0, direction="under"),
	"combined"=sapply(RD_samples,get_hpo_match_nb_exp,outlier_file=exp_outlier,HPOs_list=HPO_extended, pLI_thre=pli, direction="under"),
	institution=sapply(RD_samples,get_institution, metadata=metadata),
	affected_status=sapply(RD_samples,get_affected_status, affected_status_df=affected_status_df),
	technology=sapply(RD_samples,get_technology, metadata=metadata))


sample_outlier_filter_nogen_exp_under.df= sample_outlier_filter_nogen_exp_under.df %>% filter(raw!=0)
sample_outlier_filter_nogen_exp_under.df$prop_pLI=sample_outlier_filter_nogen_exp_under.df$raw.Lof/sample_outlier_filter_nogen_exp_under.df$raw
sample_outlier_filter_nogen_exp_under.df$prop_HPO=sample_outlier_filter_nogen_exp_under.df$raw.HPO/sample_outlier_filter_nogen_exp_under.df$raw
sample_outlier_filter_nogen_exp_under.df$prop_combined=sample_outlier_filter_nogen_exp_under.df$combined/sample_outlier_filter_nogen_exp_under.df$raw


sample_outlier_filter_nogen_exp_under.df=sample_outlier_filter_nogen_exp_under.df%>% select(sample_id,institution,affected_status, technology, prop_pLI, prop_HPO,prop_combined)

#melt DF
sample_outlier_filter_nogen_exp_under.df.m=melt(sample_outlier_filter_nogen_exp_under.df,id.vars=c("sample_id","affected_status", "institution", "technology"))

# select for only cases
sample_outlier_filter_nogen_exp_under.df.m.cases=sample_outlier_filter_nogen_exp_under.df.m %>% filter(affected_status=="Case")

# plot results
filter_withsglt.case.plot=ggplot(sample_outlier_filter_nogen_exp_under.df.m.cases, aes(x=variable, y=value, fill=variable) )+ 
	geom_boxplot(color="black", notch=T, show.legend = FALSE)+
	geom_point(size = 0, stroke = 0)+
	scale_fill_manual(values=rep("lightgrey", 3), breaks=c("prop_pLI","prop_HPO","prop_combined"),
		labels=c(expression("pLI " >="0.9"), "HPO match", "combined"), name="Filter")+
	scale_x_discrete(breaks=c("prop_pLI","prop_HPO","prop_combined"),
		labels=c("1", "2", "3"))+
	labs(x="Filter", y="Proportion of outlier genes") + RD_theme +guides(fill = guide_legend(override.aes = list(size = 4,shape =c(49,50,51))))+theme(legend.position = c(0.8,0.8))
#filter_withsglt.case.plot
ggsave('fig_2_prop_candidates_filters.pdf', filter_withsglt.case.plot, path='/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/', width=6, height=6)


save(filter_withsglt.exp.case.10kb.plot,sample_exp_outlier_filter_withsgl_10kb_up.df, sample_exp_outlier_filter_withsgl_10kb_up.df.m,filter_withsglt.exp.case.10kb_affected_status.plot,filter_withsglt.exp.case.10kb_techno.plot,file="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/exp_filters.Rdata")








#  rare variant within gene body (window method)

RV_10kb="/srv/scratch/restricted/rare_diseases/analysis/outlier_analysis/expression_level/rare_var_protein_coding_genes.txt"
RV_10kb=as.data.frame(read.table(RV_10kb, sep="\t", header=F))
colnames(RV_10kb)=c("sample_id", "gene",  "variantPos", "gnomadMAF", "raw_cadd", "phred_cadd")


# make a column with singleton included or expluded form the analysis (window method)
RV_10kb =RV_10kb %>% mutate(gnomAD_AF_withsgl = replace(gnomadMAF, gnomadMAF==".", 0))
RV_10kb =RV_10kb %>% mutate(gnomAD_AF_nosgl = replace(gnomadMAF, gnomadMAF==".", NA))


# format variant data so that only the max observed allele frequency is kept for variant were several frequencies are available (window method)
RV_10kb$gnomAD_AF_filt_withsgl=sapply(RV_10kb$gnomAD_AF_withsgl,process_variant)
RV_10kb$gnomAD_AF_filt_nosgl=sapply(RV_10kb$gnomAD_AF_nosgl,process_variant)


#filter for RV with MAF<=0.1% (window method)
maf=0.001
	# Excluding singletons
RV_nosgl_10kb=RV_10kb %>% filter(!is.na(gnomAD_AF_filt_nosgl)) %>% filter(gnomAD_AF_filt_nosgl <=maf)
	# Including singletons
RV_withsgl_10kb=RV_10kb %>% filter(!is.na(gnomAD_AF_filt_withsgl)) %>% filter(gnomAD_AF_filt_withsgl <=maf)






#neuro genes ASE


# select Cases from neurology
neuro_samples=metadata %>% filter(disease_category=="Neurology")%>% filter(affected_status=="Case")%>% select(sample_id) %>% pull
cases=metadata %>% filter(affected_status=="Case")%>% select(sample_id) %>% pull


# Read in disease gene lists
disease_genes_dir='/srv/scratch/restricted/rare_diseases/data/candidate_gene_lists/'
gene_lists=list.files(path=disease_genes_dir, pattern='genelist.txt')
disease_lists=lapply(gene_lists, read_in_disease_lists, path_to_file=disease_genes_dir)
names(disease_lists)=unlist(strsplit(gene_lists, "_genelist.txt"))

# get ensembl names for all gene lists
disease_lists_ens=lapply(disease_lists,check_gene_names)

zscore_threshold=2:6

exp_outlier_blood_neuro=exp_outlier %>% filter(sample_id %in% neuro_samples) %>% filter(sample_id %in% sample_with_data) %>%mutate(disease_gene=ifelse(gene %in% disease_lists_ens$Neurology,1,0))
exp_outlier_blood_neuro$gene=as.character(exp_outlier_blood_neuro$gene)
exp_outlier_blood_neuro$sample_id=as.character(exp_outlier_blood_neuro$sample_id)


neuro.exp.props= proportion.ratios.exp(exp_outlier_blood_neuro,zscore_threshold)
neuro.exp.props$METHOD="expression"

sig.exp <- emp_ast(neuro.exp.props$PVALUE)
neuro.exp.props$SIG=sig.exp

#neuro.exp.signed.props=  proportion.ratios.exp.signed(exp_zscores_blood_neuro,zscore_threshold)


# is ASE?
is_ase=function(sample, gene){
	#print(sample)
	#print(gene)
	if (sample %in% names(gene_ase_list)){
		ase=ifelse(gene %in% gene_ase_list[[sample]], 1, 0)
	return(ase)
}
else{return(NA)}
}
exp_outlier_blood_neuro$ase=mapply(is_ase,exp_outlier_blood_neuro$sample_id,exp_outlier_blood_neuro$gene)

exp_outlier_blood_neuro_ase=exp_outlier_blood_neuro %>% filter(ase==1)
neuro.exp.ase.props= proportion.ratios.exp(exp_outlier_blood_neuro_ase,zscore_threshold)
sig.exp.ase=  emp_ast(neuro.exp.ase.props$PVALUE)

# combine all results
neuro.prop.all=rbind(neuro.props,neuro.exp.props,neuro.combined.props)

neuro.glm.plot=ggplot(neuro.prop.all, aes(x=as.factor(THRESH), y = LOG_ODDS, color=METHOD))+
	geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH))+ theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) +  
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    #ylim(c(min(neuro.combined.props$CI.LOW) - 0.1, max(neuro.combined.props$CI.HIGH) + 0.1))+
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1, label=neuro.prop.all$SIG), size=6) +
    annotate('text', x = neuro.prop.all$THRESH-1, y =  neuro.combined.props$CI.HIGH + 0.07,
             label = formatC(neuro.combined.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme

+geom_label_repel(aes(label = SIG,
    	color = factor(METHOD)), size = 3.5, show.legend = FALSE)
disease_colors=brewer.pal(12,"Paired")

neuro.props.combined_plot.glm=ggplot(neuro.combined.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Combined')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    ylim(c(min(neuro.combined.props$CI.LOW) - 0.1, max(neuro.combined.props$CI.HIGH) + 0.1))+
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1), label=sig.combined, size=6) +

    annotate('text', x = neuro.combined.props$THRESH-1, y =  neuro.combined.props$CI.HIGH + 0.07,
             label = formatC(neuro.combined.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme
neuro.props.combined_plot.glm




neuro.exp.props_plot.glm=ggplot(neuro.exp.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Expression Outlier')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1), label=sig.exp, size=6) +

    ylim(c(min(neuro.exp.props$CI.LOW) - 0.1, max(neuro.exp.props$CI.HIGH) + 0.1))+
    annotate('text', x = neuro.exp.props$THRESH-1, y =  neuro.exp.props$CI.HIGH + 0.07,
             label = formatC(neuro.exp.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme

neuro.exp.ase.props_plot.glm=ggplot(neuro.exp.ase.props, aes(x = as.factor(THRESH), y = LOG_ODDS)) +
    geom_pointrange(aes(ymin = CI.LOW, ymax = CI.HIGH), color=disease_colors[1]) + theme_bw() +
    xlab('Z-score threshold') + ylab('Log Odds Ratio') + geom_hline(yintercept = 0) + ggtitle('Expression Outlier')+ 
    theme(legend.key = element_blank(),
          legend.position = c(0.8,0.65),
          legend.background = element_blank()) +
    geom_text(aes(x=as.factor(THRESH), y=CI.HIGH+0.1), label=sig.exp.ase, size=6) +

    ylim(c(min(neuro.exp.ase.props$CI.LOW) - 0.1, max(neuro.exp.ase.props$CI.HIGH) + 0.1))+
    annotate('text', x = neuro.exp.ase.props$THRESH-1, y =  neuro.exp.ase.props$CI.HIGH + 0.07,
             label = formatC(neuro.exp.ase.props$COUNT, format = "d", big.mark = ","), size = 3.5)+ RD_theme


## combined_plot
sup_ase_neuro_plot=ggdraw()+draw_plot(neuro.exp.props_plot.glm, 0,0,1/2,1)+
	draw_plot(neuro.exp.ase.props_plot.glm, 1/2,0,1/2,1)+
	draw_plot_label(c('A', 'B'), c(0,1/2), c(1,1), size = 20)


pdf('/srv/scratch/restricted/rare_diseases/analysis/manuscript/figures/sup_ase_neuro_plot.pdf', w=11, h=8)
sup_ase_neuro_plot
dev.off()

