# =======================================================================================================
# This is a script for generating EDF1: Gene expression patterns across whole blood samples
#
# Note that the final figure was generated by using inkscape by combining figures from this script for visualization purposes
#
# =======================================================================================================



#!/bin/R


# load required functions
source('Figures_source.R')



#  Read in input data
edf1a=read.table("EDF1a_data.txt", sep="\t", header=T)
edf1b=read.table("EDF1b_data.txt", sep="\t", header=T)
edf1c=read.table("EDF1c_data.txt", sep="\t", header=T)
edf1d=read.table("EDF1d_data.txt", sep="\t", header=T)


#--- Libraries

library(cowplot)
library(ggplot2)
library(RColorBrewer)


fsize=15
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize),
	axis.text.y= element_text(size=fsize), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.5),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=fsize),
	panel.border=element_blank()) 


# Panel A
edf1a_plot=ggplot(edf1a, aes(x=percent_covered))+geom_density(fill="gray")+
	geom_vline(aes(xintercept=mean(percent_covered)), color="blue",linetype="dashed")+
	geom_vline(aes(xintercept=median(percent_covered)), color="red",linetype="dashed")+
	labs(x="Proportion of annotated junctions covered per gene", y="Density")+
	RD_theme


# Panel B
disease_colors=brewer.pal(12,"Paired")
col_juncplot=c(grey.colors(n=11,start = 0.6, end = 1)[1:4],disease_colors[10])
edf1b$variable=factor(edf1b$variable, levels=c("Hematology", "Ophthalmology","Musculoskeletal", "Neurology", "OMIM"))

edf1b_plot=ggplot(edf1b, aes(x=variable, y=value, fill=variable))+ 
	geom_bar(stat="identity")+ 
	scale_fill_manual(values=col_juncplot)+
	labs(x="", y="% Genes for which at least one junction\ncovered with at least 5 reads in 20% samples")+
	RD_theme+
	coord_flip()+ 
	theme(legend.position="")


# Panel C

col_1C_grey=c("#C0BCB6",  disease_colors[10])

edf1c_plot=ggplot(edf1c, aes(x=variable, y=value, fill=multi)) +
	geom_boxplot(outlier.colour="black", notch=TRUE)+
	RD_theme +
	labs(x='', y='Value')+
	scale_fill_manual(values=col_1C_grey,name="Expression", labels=c("Single tissue\nGTEx v7 (N=1)","Multiple tissues\nGTEx v7(N>1)"))+
	theme(legend.position=c(0.2,0.7))+
	ylim(-15, 25)+ylab("Z-score")+xlab("Variant category")+
	scale_x_discrete(labels=c("Synonymous", "Missense", "LoF"))#+stat_compare_means(aes(group = multi,label = paste0("p = ", ..p.format..), size=fsize))

## Panel D

edf1d_plot=ggplot(edf1d, aes(x=bin))+
	geom_bar( width=0.4,fill= disease_colors[10], position = position_dodge(width = 0.1), color="black")+ 
	geom_text(stat='count',aes(label=..count..),vjust=1.3, color="white", size=fsize/3)+
	RD_theme +  scale_x_discrete(labels=c("[0,0.1)" = "[0,0.1)", "[0.1,10)" = "[0.1,10)", "[10,1e+04)" = "[10,1e+05)"))+
	labs(x= "Average TPM", y="Number of LoF intolerant genes")
	#coord_flip()


# Make combined plot
edf1_plot=ggdraw()+
	draw_plot(edf1a_plot, 0,1/2,1/2,1/2)+
	draw_plot(edf1b_plot, 1/2,1/2,1/2,1/2)+
	draw_plot(edf1c_plot, 0,0,1/2,1/2)+
	draw_plot(edf1d_plot, 1/2,0,1/2,1/2)+
	draw_plot_label(c('A', 'B','C','D'), c(0,1/2,0,1/2), c(1,1,1/2,1/2), size = 15)

