# =======================================================================================================
# This is a script for generating FigureS2: Junctions coverage across whole blood samples
# #
# input:
#       1. FigureS2.in.RData
#          Data generated by running Figure_S2_getdata.R
#
# output:
#       1. FigureS2A.pdf, FigureS2B.pdf, FigureS2.pdf
#       2. FigureS2.out.RData
#          Store all data used and generated by running this script
#
# Note that the final figure was generated by using inkscape by combining figures from this script for visualization purposes
#
# =======================================================================================================



#!/bin/R

# look at the number of reads covering junctions for genes of interest

# Master directory
dir = Sys.getenv('RARE_DIS_DIR')


# load required functions
source('Figures_source.R')



#  Load input data
load(file = paste0(dir,"/analysis/manuscript/figures_revision/FigureS2.in.RData"))



#--- Libraries

library(cowplot)
library(ggplot2)
library(RColorBrewer)


fsize=15
RD_theme=theme_classic()+
	theme(axis.text.x= element_text(size=fsize),
	axis.text.y= element_text(size=fsize), 
	axis.title = element_text(size = fsize), 
    legend.text = element_text(size = fsize-1), 
    legend.title = element_text(size = fsize), 
    axis.ticks = element_line(size = 0.5),
	strip.background=element_blank(),
	panel.grid.major=element_blank(),
	panel.grid.minor=element_blank(),
	axis.text=element_text(size=fsize),
	panel.border=element_blank()) 


# Panel A
junction_coverage_density_plot=ggplot(nb_junction_cov, aes(x=percent_covered))+geom_density(fill="gray")+
	geom_vline(aes(xintercept=mean(percent_covered)), color="blue",linetype="dashed")+
	geom_vline(aes(xintercept=median(percent_covered)), color="red",linetype="dashed")+
	labs(x="Proportion of annotated junctions covered per gene", y="Density")+
	RD_theme

ggsave('FigureS2C.pdf', junction_coverage_density_plot, path=paste(dir,"/analysis/manuscript/figures_revision/", sep=""), width=6, height=6)

# Panel B
disease_colors=brewer.pal(12,"Paired")
col_juncplot=c(grey.colors(n=11,start = 0.6, end = 1)[1:4],disease_colors[10])

disease_junc_plot=ggplot(disease_junc.m, aes(x=variable, y=value, fill=variable))+ 
	geom_bar(stat="identity")+ 
	scale_fill_manual(values=col_juncplot)+
	labs(x="", y="% Genes for which at least one junction\ncovered with at least 5 reads in 20% samples")+
	RD_theme+
	coord_flip()+ 
	theme(legend.position="")
disease_junc_plot
ggsave('FigureS2D.pdf', disease_junc_plot,path=paste(dir,"/analysis/manuscript/figures_revision/", sep=""), width=7, height=6)





# Figure 1C
col_1C_grey=c("#C0BCB6",  disease_colors[10])
exp_multtissues_genes_plot_greyscale=ggplot(exp_multtissues_genes.m, aes(x=variable, y=value, fill=multi)) +
	geom_boxplot(outlier.colour="black", notch=TRUE)+
	RD_theme +
	labs(x='', y='Value')+
	scale_fill_manual(values=col_1C_grey,name="Expression", labels=c("Single tissue\nGTEx v7 (N=1)","Multiple tissues\nGTEx v7(N>1)"))+
	theme(legend.position=c(0.2,0.7))+
	ylim(-15, 25)+ylab("Z-score")+xlab("Variant category")+
	scale_x_discrete(labels=c("Synonymous", "Missense", "LoF"))+stat_compare_means(aes(group = multi,label = paste0("p = ", ..p.format..), size=fsize))

exp_multtissues_genes_plot_greyscale
ggsave('FigureS2A.pdf', exp_multtissues_genes_plot_greyscale, path=paste(dir,"/analysis/manuscript/figures_revision/", sep=""), width=6, height=6)


## Figure 1D 

exp_blood_pLI_plot=ggplot(tpms_df_avg, aes(x=bin))+
	geom_bar( width=0.4,fill= disease_colors[10], position = position_dodge(width = 0.1), color="black")+ 
	geom_text(stat='count',aes(label=..count..),vjust=1.3, color="white", size=fsize/3)+
	RD_theme +  scale_x_discrete(labels=c("[0,0.1)" = "[0,0.1)", "[0.1,10)" = "[0.1,10)", "[10,1e+04)" = "[10,1e+05)"))+
	labs(x= "Average TPM", y="Number of LoF intolerant genes")
	#coord_flip()
exp_blood_pLI_plot

ggsave('FigureS2B.pdf', exp_blood_pLI_plot, path=paste(dir,"/analysis/manuscript/figures_revision/", sep=""), width=6, height=6)


# Make combined plot
combined_plots=ggdraw()+
	draw_plot(junction_coverage_density_plot, 0,1/2,1/2,1/2)+
	draw_plot(disease_junc_plot, 1/2,1/2,1/2,1/2)+
	draw_plot(exp_multtissues_genes_plot_greyscale, 0,0,1/2,1/2)+
	draw_plot(exp_blood_pLI_plot, 1/2,0,1/2,1/2)+
	draw_plot_label(c('A', 'B','C','D'), c(0,1/2,0,1/2), c(1,1,1/2,1/2), size = 15)

#combined_plots
pdf(paste(dir,"/analysis/manuscript/figures_revision/FigureS2.pdf",sep=""), w=9, h=9)
combined_plots
dev.off()

# Save data
save.image(file = paste0(dir,"/analysis/manuscript/figures_revision/FigureS2.out.RData"))
